{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2c0bb48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # A00_environment_and_genome_setup\n",
    "\n",
    "# # in terminal, set up snmCT pipeline environment\n",
    "# module load anaconda3 # or otherwise activate conda\n",
    "# conda env create -f Documentation/snmCTseq.yml\n",
    "\n",
    "# # optional: at this stage, i typically create a project-specific downstream analysis\n",
    "# # environment as well (e.g., scverse/scanpy, etc) for reproducibility/working in juypter, e.g.,\n",
    "# # conda env create --name IGVF_mCT # followed by conda/mamba pkg installs, incl ipykernel\n",
    "# # conda env export | grep -v \"^prefix: \" > conda-environment.yml # export versions\n",
    "# # python -m ipykernel install --user --name IGVF_mCT --display-name \"IGVF\" # juypter kernel\n",
    "\n",
    "# # genome setup overall commands\n",
    "# # ‡ fast enough to run interactively\n",
    "\n",
    "# qsub Scripts/A00a_genome_dl_index.sub  # ‡\n",
    "# qsub Scripts/A00b_genome_prep_bismark.sub\n",
    "# qsub Scripts/A00c_genome_prep_star.sub\n",
    "# qsub Scripts/A00d_gtf_annotations_bed.sub # ‡"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6978ac0d",
   "metadata": {},
   "source": [
    "## minimal project directory setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c092f422",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # run in terminal\n",
    "# dir_proj=/u/project/cluo/chliu/Analyses/IGVF\n",
    "# mkdir $dir_proj; cd $dir_proj\n",
    "# mkdir fastq_demultip fastq_raw fastq_trimmed mapping_bismark mapping_star\n",
    "# mkdir Metadata Notebooks Scripts sublogs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8225a9d",
   "metadata": {},
   "source": [
    "## snmCT_parameters (crucial to review)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23fe91be",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../snmCT_parameters.env\n",
    "\n",
    "# parameters file --------------------------------------------------------------\n",
    "# note: filepaths are relative to project directory or absolute paths\n",
    "# recommend using absolute filepaths for all \"ref_dir\"/genome related files by\n",
    "# find & replace \"/u/project/cluo/chliu/Genomes/human_gencode_v40\" -->\n",
    "#        folder where your genome assembly resides\n",
    "\n",
    "# primary analysis/project folder\n",
    "dir_proj=/u/project/cluo/chliu/Analyses/IGVF\n",
    "\n",
    "# scratch folder if available, otherwise can set as $dir_proj\n",
    "dir_scratch=/u/project/cluo_scratch/chliu\n",
    "\n",
    "# folder with raw data\n",
    "# (.fastq, for our group usually split across 4 lanes of a novaseq run)\n",
    "dir_originalfastq=/u/project/cluo/Shared_Datasets/IGVF/202208_Pilot/snmCT-seq/fastq/\n",
    "  \n",
    "# reference genome/files -----------------------------------------------------\n",
    "# if a Hoffman2 user with access to our partition, can use the below hg38 paths\n",
    "# (versions also exist +/- lambda spike-in)\n",
    "ref_dir=/u/project/cluo/chliu/Genomes/human_gencode_v40\n",
    "ref_fasta=/u/project/cluo/chliu/Genomes/human_gencode_v40/GRCh38.primary_assembly.genome.fa\n",
    "ref_gtf=/u/project/cluo/chliu/Genomes/human_gencode_v40/gencode.v40.primary_assembly.annotation.gtf\n",
    "\n",
    "# # (or this directory in commented code for mouse)\n",
    "# ref_dir=/u/project/cluo/chliu/Genomes/mouse_gencode_v29\n",
    "# ref_fasta=/u/project/cluo/chliu/Genomes/mouse_gencode_v29/GRCm39.primary_assembly.genome.fa\n",
    "# ref_gtf=/u/project/cluo/chliu/Genomes/mouse_gencode_v29/gencode.vM29.primary_assembly.annotation.gtf\n",
    "\n",
    "\n",
    "\n",
    "# parameters that usually do not have to be changed ----------------------------\n",
    "# (besides organism name; e.g., replace human with mouse_gencode_v29)\n",
    "\n",
    "# reference genome\n",
    "ref_chromsizes=/u/project/cluo/chliu/Genomes/human_gencode_v40/chromsizes.tsv\n",
    "ref_genebody=/u/project/cluo/chliu/Genomes/human_gencode_v40/annotations/genebody.tsv\n",
    "ref_geneslop2k=/u/project/cluo/chliu/Genomes/human_gencode_v40/annotations/geneslop2k.tsv\n",
    "ref_exon=/u/project/cluo/chliu/Genomes/human_gencode_v40/annotations/exon.tsv\n",
    "ref_flat=/u/project/cluo/chliu/Genomes/human_gencode_v40/annotations/refFlat.txt.gz\n",
    "ref_rrna=/u/project/cluo/chliu/Genomes/human_gencode_v40/annotations/rRNA.intervallist\n",
    "ref_starfolder=/u/project/cluo/chliu/Genomes/human_gencode_v40/STAR\n",
    "\n",
    "# metadata files\n",
    "metadat_plate=Metadata/A01b_plate_metadata.csv\n",
    "metadat_well=Metadata/A01c_well_filepath.csv\n",
    "\n",
    "# fastqc\n",
    "wells_to_run=Scripts/A02c_random_fastqc_wells.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52e88a95",
   "metadata": {},
   "source": [
    "## (A00a) download reference .fasta, index, get chrom sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0778a67b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # note: if using UCLA Hoffman2 cluster, can use my reference genomes\n",
    "# # (see filepaths listed above)\n",
    "\n",
    "# # there may be extraneous files in my refs from genotyping pipelines NOT needed for mCT\n",
    "# # (.fa.gz --> .bwt, .pac, .ann, .amb, .sa, .fai, .dict)\n",
    "# bwa index GRCh38.primary_assembly.genome.fa.gz\n",
    "# gatk CreateSequenceDictionary -R GRCh38.primary_assembly.genome.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af0d41f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A00a_genome_dl_index.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A00a_genome_dl_index.$JOB_ID\n",
    "#$ -j y\n",
    "#$ -N A00a_genome_dl_index\n",
    "#$ -l h_rt=4:00:00,h_data=8G\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init ------------------—------------------—-----------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snmCTseq # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "\n",
    "\n",
    "# download/extract ref ------------------—------------------—-------------------\n",
    "\n",
    "cd ${ref_dir}\n",
    "\n",
    "# download gencode v40 (GRCh38.p13) .fa.gz & .gtf.gz as example\n",
    "# double check filepaths and hard-coded names here \n",
    "wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_40/GRCh38.primary_assembly.genome.fa.gz\n",
    "wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_40/gencode.v40.primary_assembly.annotation.gtf.gz\n",
    "\n",
    "# extract compressed files as needed\n",
    "# put final, uncompressed names here into \"snmCT_parameters.env\" (as \"ref_fasta\" and \"ref_gtf\")\n",
    "for gzfile in *gz\n",
    "do\n",
    "    gunzip -c ${gzfile} > ${gzfile/.gz/}\n",
    "done\n",
    "\n",
    "# note: if ref_fasta doesn't have .fa or .fasta file format\n",
    "# bismark may have issues detecting it in ${ref_dir}\n",
    "# rename .fna --> .fa, for example\n",
    "# for f in *fna; do mv $f ${f/fna/fa}; done\n",
    "\n",
    "\n",
    "# note on working with Lambda phage or other spike-in (bisulfite conversion efficiency QC)\n",
    "# nice to add to reference genome at this step\n",
    "# e.g., Escherichia phage Lambda, complete genome (GenBank: J02459.1) via NCBI\n",
    "# wget \"https://www.ncbi.nlm.nih.gov/sviewer/viewer.cgi?tool=portal&save=file$=seqview&db=nuccore&report=fasta&id=215104\" -O lambda.fa\n",
    "# sed -i \"1s/.*/>chrL/\" lambda.fa\n",
    "# cat GRCh38.primary_assembly.genome.fa lambda.fa > GRCh38_plus_lambda.fa \n",
    "\n",
    "\n",
    "\n",
    "# index, chrom sizes -----------------—------------------—----------------------\n",
    "\n",
    "# extract bp length/chromosome\n",
    "samtools faidx ${ref_fasta}\n",
    "cut -f 1-2 ${ref_fasta}.fai > chromsizes.tsv\n",
    "\n",
    "# .fa --> .dict\n",
    "picard CreateSequenceDictionary -R ${ref_fasta}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A00a_bwa_index' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID ended on:   \" `date `\n",
    "echo \" \""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "016fe42d",
   "metadata": {},
   "source": [
    "## (A00b) prep bismark genome ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b0ee576",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A00b_genome_prep_bismark.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A00b_genome_prep_bismark.$JOB_ID\n",
    "#$ -j y\n",
    "#$ -N A00b_prep_bismark\n",
    "#$ -l h_rt=24:00:00,h_data=8G,highp\n",
    "#$ -pe shared 4\n",
    "#$ -hold_jid A00a_genome_dl_index\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init ------------------—------------------—-----------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snmCTseq # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "\n",
    "\n",
    "# bismark index ------------------—------------------—--------------------------\n",
    "\n",
    "cd ${ref_dir}\n",
    "\n",
    "# note: if ref_fasta doesn't have .fa or .fasta file format\n",
    "# bismark may have issues detecting it in ${ref_dir}\n",
    "bismark_genome_preparation ${ref_dir} --bowtie2\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A00b_genome_prep_bismark' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID ended on:   \" `date `\n",
    "echo \" \""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5b1f0bc",
   "metadata": {},
   "source": [
    "## (A00c) prep STAR genome ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6b9de87",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A00c_genome_prep_star.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A00c_genome_prep_star.$JOB_ID\n",
    "#$ -j y\n",
    "#$ -N A00c_prep_star\n",
    "#$ -l h_rt=24:00:00,h_data=8G,highp\n",
    "#$ -pe shared 8\n",
    "#$ -hold_jid A00a_genome_dl_index\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "# environment init ------------------—------------------—-----------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snmCTseq # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "\n",
    "\n",
    "# star index ------------------—------------------—-----------------------------\n",
    "# assumes 2x150 paired-end\n",
    "\n",
    "STAR --runThreadN 8 \\\n",
    "--runMode genomeGenerate \\\n",
    "--genomeDir ${ref_dir}/STAR149 \\\n",
    "--genomeFastaFiles ${ref_fasta} \\\n",
    "--sjdbGTFfile ${ref_gtf} \\\n",
    "--sjdbOverhang 149\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A00c_genome_prep_star' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID ended on:   \" `date `\n",
    "echo \" \""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6ea0359",
   "metadata": {},
   "source": [
    "## (A00d) extract annotations from gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "699c6112",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A00d_gtf_annotations_bed.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A00d_gtf_annotations_bed.$JOB_ID\n",
    "#$ -j y\n",
    "#$ -N A00d_gtf_annotations_bed\n",
    "#$ -l h_rt=0:30:00,h_data=8G\n",
    "#$ -hold_jid A00a_genome_dl_index\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "# environment init ------------------—------------------—-----------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snmCTseq # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "\n",
    "\n",
    "# process .gtf --> .bed ------------------—------------------—------------------\n",
    "\n",
    "mkdir ${ref_dir}/annotations\n",
    "\n",
    "python Scripts/A00d_gtf_annotations_bed.py\n",
    "\n",
    "\n",
    "\n",
    "# UCSC/Picard utilities for QC metrics ------------------—------------------—---\n",
    "\n",
    "cd ${ref_dir}\n",
    "\n",
    "# rRNA interval list\n",
    "picard BedToIntervalList -I annotations/rRNA.bed -O \\\n",
    "annotations/rRNA.intervallist -SD ${ref_fasta/fa/dict}\n",
    "\n",
    "\n",
    "# rRNA interval list\n",
    "picard BedToIntervalList -I annotations/rRNA.bed -O \\\n",
    "    annotations/rRNA.intervallist -SD ${ref_fasta/fa/dict}\n",
    "\n",
    "# ref flat: equiv to genepred fmt minus some tailing columns\n",
    "gtfToGenePred -genePredExt -geneNameAsName2 ${ref_gtf} refFlat.tmp.txt\n",
    "cut -f 12 refFlat.tmp.txt > refFlat.tmp1\n",
    "cut -f 1-10 refFlat.tmp.txt > refFlat.tmp2\n",
    "\n",
    "paste refFlat.tmp1 refFlat.tmp2 > annotations/refFlat.txt\n",
    "gzip annotations/refFlat.txt\n",
    "\n",
    "rm refFlat.tmp*\n",
    "  \n",
    "\n",
    "  \n",
    "  \n",
    "\n",
    "echo -e \"\\n\\n'A00d_gtf_annotations_bed' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID ended on:   \" `date `\n",
    "echo \" \""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e65bab1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A00d_gtf_annotations_bed.py\n",
    "\n",
    "# ==============================================================================\n",
    "# A00d_gtf_annotations_bed.py \n",
    "# exports four annotation-related files to $ref_dir (reference genome)\n",
    "# for mcds creation & down-stream analysis\n",
    "# ==============================================================================\n",
    "\n",
    "# recommend running interactively in python/Jupyter to check outputs\n",
    "# works for .gtfs from GENCODE, but double check fmt for other sources\n",
    "\n",
    "# if running interactively, check snmCT_parameters.env loaded or manually spec os.environ e.g.,\n",
    "# os.environ['ref_dir'] = \"/u/project/cluo/chliu/Genomes/human_gencode_v40\"\n",
    "# os.environ['ref_gtf'] = \"/u/project/cluo/chliu/Genomes/human_gencode_v40/gencode.v40.primary_assembly.annotation.gtf\"\n",
    "# os.environ['ref_chromsizes'] = \"/u/project/cluo/chliu/Genomes/human_gencode_v40/chromsizes.tsv\"\n",
    "\n",
    "\n",
    "\n",
    "# load packages ----------------------------------------------------------------\n",
    "\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "\n",
    "\n",
    "# load reference info ----------------------------------------------------------\n",
    "\n",
    "os.chdir(os.environ['ref_dir'])\n",
    "os.makedirs(\"annotations/\", exist_ok=True)\n",
    "\n",
    "gtf_file = pd.read_csv(os.environ['ref_gtf'],\n",
    "                       comment = \"#\", delimiter=\"\\t\", header = None)\n",
    "chrom_sizes = pd.read_csv(os.environ['ref_chromsizes'], sep = \"\\t\", header = None)\n",
    "chrom_sizes.columns = ['#chr', 'chrlen'] \n",
    "\n",
    "\n",
    "\n",
    "# genebody ---------------------------------------------------------------------\n",
    "# .gtf to .bed (1-based --> 0 based)\n",
    "# columns: chr, start, end, ENSG identifier\n",
    "# note that this contains mitochondrial, ribosomal, lncRNAs, etc;\n",
    "# this may or may not be desireable in downstream analyses\n",
    "\n",
    "genebody = gtf_file[gtf_file.iloc[:, 2] == 'gene'].iloc[:, [0, 3, 4, 8]]\n",
    "genebody.iloc[:, 1] = genebody.iloc[:, 1] - 1 # start changes to 0-pos\n",
    "genebody.columns = ['#chr', 'start', 'end', 'annot']\n",
    "genebody.reset_index(inplace=True, drop=True)\n",
    "\n",
    "# extract info from the annot column (;, \")\n",
    "genebody['gene'] = genebody['annot'].transform(lambda x: str(x).split('\\\"')[1])\n",
    "genebody['symbol'] = genebody['annot'].transform(lambda x: str(x).split('\\\"')[5])\n",
    "genebody['type'] = genebody['annot'].transform(lambda x: str(x).split('\\\"')[3])\n",
    "\n",
    "# .gtf checks: should be zero\n",
    "if sum(genebody.gene.duplicated()) != 0 | sum(genebody.start >= genebody.end) != 0:\n",
    "  print(\"WARNING: check .bed outputs; was gene info was processed correctly from .gtf?\")\n",
    "\n",
    "# export ENSG --> Symbol\n",
    "genebody.drop('annot', axis = 1, inplace = True)\n",
    "genebody.to_csv(\"annotations/ensembl_to_symbol.tsv\", sep = \"\\t\", index = False)\n",
    "\n",
    "# bed4 format for .allcools\n",
    "genebody = genebody.iloc[:, 0:4]\n",
    "genebody.to_csv(\"annotations/genebody.bed\", sep = \"\\t\", index = False)\n",
    "\n",
    "\n",
    "\n",
    "# gene +/- 2kb -----------------------------------------------------------------\n",
    "# above, but padding a 2kb region\n",
    "# past manuscripts do this for higher mC modality coverage, but less interpretable\n",
    "\n",
    "g2k = genebody.copy()\n",
    "g2k.iloc[:, 1] = g2k.iloc[:, 1] - 2000\n",
    "g2k.iloc[:, 2] = g2k.iloc[:, 2] + 2000\n",
    "\n",
    "# but check +/-2kb still within chromosome length\n",
    "# (low # genes affected, but may cause downstream issues)\n",
    "g2k.loc[g2k.start < 0, 'start'] = 0\n",
    "\n",
    "g2k = pd.merge(g2k, chrom_sizes, on = '#chr') \n",
    "filter_chrlen = g2k.end > g2k.chrlen\n",
    "g2k.loc[filter_chrlen, 'end'] = g2k.chrlen[filter_chrlen]\n",
    "g2k.drop('chrlen', axis = 1, inplace=True)\n",
    "g2k.to_csv(\"annotations/geneslop2k.bed\", sep = \"\\t\", index = False)\n",
    "\n",
    "\n",
    "# exonic -----------------------------------------------------------------------\n",
    "# exon-level annotations, useful for STAR exon-only quant, potential ASE\n",
    "\n",
    "exon = gtf_file[gtf_file.iloc[:, 2] == 'exon'].iloc[:, [0, 3, 4, 8]]\n",
    "exon.iloc[:, 1] = exon.iloc[:, 1] - 1 # start changes to 0-pos\n",
    "exon.columns = ['#chr', 'start', 'end', 'annot']\n",
    "\n",
    "exon['gene'] = exon['annot'].transform(lambda x: str(x).split('\\\"')[1])\n",
    "exon['transcript'] =  exon['annot'].transform(lambda x: str(x).split('\\\"')[3])\n",
    "exon = exon.drop('annot', axis = 1).reset_index(drop=True)\n",
    "exon.to_csv(\"annotations/exon.bed\", sep = \"\\t\", index = False)\n",
    "\n",
    "\n",
    "\n",
    "# rRNA genes -------------------------------------------------------------------\n",
    "# for some QC metrics after RNA alignments\n",
    "\n",
    "rRNA = genebody.loc[gtf_file.iloc[:, 8].str.contains(\"rRNA\"), :]\n",
    "rRNA.to_csv(\"annotations/rRNA.bed\", sep = \"\\t\", index = False)\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
