{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## A02_demultiplexing overall commands\n",
    "\n",
    "# qsub Scripts/A02a_demultiplex_fastq.sub # *\n",
    "# qsub Scripts/A02b_check_demultip.sub # ‡\n",
    "# qsub Scripts/A02c_fastqc_demultip_fastq.sub\n",
    "\n",
    "# # * = job array based on \"platenum\"\n",
    "# # † = job array based on \"batchnum\" (two rows at a time)\n",
    "# # ‡ fast enough to run interactively"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (A02a) demultiplex .fastq.gz files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A02a_demultiplex_fastq.pl\n",
    "\n",
    "#!/usr/bin/perl -w\n",
    "use strict;\n",
    "\n",
    "# A02a_demultiplex_fastq.pl, v0.1 ==============================================\n",
    "# perl script originally written by Dr. Chongyuan Luo (@luogenomics)\n",
    "# minor modifications for readability/documentation by Choo Liu (@chooliu) \n",
    "# inputs: - R1 and R2 .fastq.gz file in 'fastq_raw'\n",
    "#           containing all cells, first 8bp of Read 1 assumed to be cell barcode\n",
    "#         - index .fasta (> 384-well plate location / cell barcode sequence)\n",
    "# outputs: - R1 and R2 .fastq.gz for each cell\n",
    "#          - summary .txt showing # of reads per cell barcode\n",
    "# typical usage: perl A02a_demultiplex_fastq.pl r1.fq.gz r2.fq.gz index.fa\n",
    "# ==============================================================================\n",
    "\n",
    "#!/usr/bin/perl -w\n",
    "use strict;\n",
    "\n",
    "my %index_seq; my %index; my $index_seq; my $index_str; my @index=();\n",
    "open index_file, \"$ARGV[2]\" or die $!;\n",
    "while (<index_file>)\n",
    "{\n",
    "  chop $_; $index_str=substr($_,1,length($_)-1);\n",
    "  $_=<index_file>; \n",
    "  chop $_; $index_seq{substr($_,1,length($_)-1)}=$index_str; $index{$index_str}=substr($_,1,length($_)-1);\n",
    "  push (@index,$index_str);\n",
    "}\t\n",
    "close index_file;\n",
    "\n",
    "my $l_fastq=@ARGV;\n",
    "my @r1_list; my @r2_list; my @r1_name; my @r2_name; my $r1_name; my $r2_name;\n",
    "if ($l_fastq==2) { push (@r1_list,$ARGV[0]); }\n",
    "elsif ($l_fastq==3) { push (@r1_list,$ARGV[0]); push (@r2_list,$ARGV[1]); }\n",
    "\n",
    "\n",
    "my $count; my %index_count; my $eight; my @r1_tmp; my @r2_tmp; \n",
    "my $r1; my $r2; my $percentage; my $command; my $return;\n",
    "\n",
    "\n",
    "for (my $sample=0; $sample<=@r1_list-1; $sample++)\n",
    "{\n",
    "  if (@r2_list>0)\n",
    "  {\n",
    "    @r1_name=split(/_R1/,$r1_list[$sample]);\n",
    "    @r2_name=split(/_R2/,$r2_list[$sample]);\n",
    "    \n",
    "    my %r1_out; my %r2_out;\n",
    "    for (my $index=0; $index<=@index-1; $index++)\n",
    "    {\n",
    "      $r1_name[0] =~ s/fastq_raw/fastq_demultip/ig;\n",
    "      $r2_name[0] =~ s/fastq_raw/fastq_demultip/ig;\n",
    "      local *FILE;\n",
    "      open FILE, \" | gzip -c > $r1_name[0]\\_$index[$index]\\_indexed_R1$r1_name[1]\" or die $!;\n",
    "      $r1_out{$index[$index]}=*FILE;\n",
    "      local *FILE;\n",
    "      open FILE, \" | gzip -c > $r2_name[0]\\_$index[$index]\\_indexed_R2$r2_name[1]\" or die $!;\n",
    "      $r2_out{$index[$index]}=*FILE;\n",
    "    }\t\n",
    "    \n",
    "    my $count; my %index_count; my $six;\n",
    "    open r1_fastq_in, \"gzip -cd $r1_list[$sample] | \" or die $!;\n",
    "    open r2_fastq_in, \"gzip -cd $r2_list[$sample] | \" or die $!;\n",
    "    $count=0; @r1_tmp=(); @r2_tmp=();\n",
    "    while (<r1_fastq_in>)\n",
    "    {\n",
    "      chop $_; $count++;\n",
    "      $r1=$_; $r2=<r2_fastq_in>; chop $r2;\n",
    "      push (@r1_tmp, $r1); push (@r2_tmp,$r2);\n",
    "      if (!($count%4))\n",
    "      {\n",
    "        $eight=substr($r1_tmp[1],0,8);\n",
    "        if (exists $index_seq{$eight})\n",
    "        {\n",
    "          if (!exists $index_count{$index_seq{$eight}}) { $index_count{$index_seq{$eight}}=0; }\n",
    "          $index_count{$index_seq{$eight}}++;\n",
    "          print {$r1_out{$index_seq{$eight}}} \"$r1_tmp[0]\\n$r1_tmp[1]\\n$r1_tmp[2]\\n$r1_tmp[3]\\n\";\n",
    "          print {$r2_out{$index_seq{$eight}}} \"$r2_tmp[0]\\n$r2_tmp[1]\\n$r2_tmp[2]\\n$r2_tmp[3]\\n\";\n",
    "        }\n",
    "        else\n",
    "        {\n",
    "          if (!exists $index_count{0}) { $index_count{0}=0; }\n",
    "          $index_count{0}++;\n",
    "        }\n",
    "        @r1_tmp=(); @r2_tmp=();\n",
    "      }\n",
    "    }\n",
    "    for (my $index=0; $index<=@index-1; $index++)\n",
    "    {\n",
    "      close $r1_out{$index[$index]};\n",
    "      close $r2_out{$index[$index]};\n",
    "      if (!exists $index_count{$index[$index]}) \n",
    "      {\n",
    "        $command=\"rm $r1_name[0]\\_$index[$index]\\_indexed_R1$r1_name[1]\";\n",
    "        print \"$command\\n\"; $return=system($command);\n",
    "        $command=\"rm $r2_name[0]\\_$index[$index]\\_indexed_R2$r2_name[1]\";\n",
    "        print \"$command\\n\"; $return=system($command);\n",
    "      }\n",
    "    }\n",
    "    close r1_fastq_in;\n",
    "    close r2_fastq_in;\n",
    "    my $indexout = substr($ARGV[2], -4, 1);\n",
    "    open summary_out, \">$r1_name[0]\\_summary\\_$indexout.txt\" or die $!;\n",
    "    $count=int($count/4);\n",
    "    print summary_out \"total reads - $count\\n\";\n",
    "    for (sort {$a cmp $b} keys %index_count) \n",
    "    {\n",
    "      $percentage=sprintf('%.1f', 100*($index_count{$_}/$count)); $percentage=\"$percentage%\";\n",
    "      if (!$_) { print summary_out \"undetermined index\\t\\t$index_count{$_}\\t$percentage\\n\"; }\n",
    "      else { print summary_out \"index $_\\t$index{$_}\\t$index_count{$_}\\t$percentage\\n\"; }\n",
    "    }\n",
    "    close summary_out;\n",
    "  }\n",
    "} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A02a_cellbarcodes_subset1.fa\n",
    ">A1\n",
    "^ACGATCAG\n",
    ">A3\n",
    "^TCGAGAGT\n",
    ">A5\n",
    "^CTAGCTCA\n",
    ">A7\n",
    "^ATCGTCTC\n",
    ">A9\n",
    "^TCGACAAG\n",
    ">A11\n",
    "^CCTTGGAA\n",
    ">A13\n",
    "^ATCATGCG\n",
    ">A15\n",
    "^TGTTCCGT\n",
    ">A17\n",
    "^ATTAGCCG\n",
    ">A19\n",
    "^CGATCGAT\n",
    ">A21\n",
    "^GATCTTGC\n",
    ">A23\n",
    "^AGGATAGC\n",
    ">C1\n",
    "^GTAGCGTA\n",
    ">C3\n",
    "^AGAGTCCA\n",
    ">C5\n",
    "^GCTACTCT\n",
    ">C7\n",
    "^CTCTGGAT\n",
    ">C9\n",
    "^AGATCGTC\n",
    ">C11\n",
    "^GCTCAGTT\n",
    ">C13\n",
    "^GTCCTAAG\n",
    ">C15\n",
    "^TATGGCAC\n",
    ">C17\n",
    "^TCGGATTC\n",
    ">C19\n",
    "^AACAGCGA\n",
    ">C21\n",
    "^CCAACGAA\n",
    ">C23\n",
    "^CAGTGCTT\n",
    ">E1\n",
    "^GATCAAGG\n",
    ">E3\n",
    "^TCTTCGAC\n",
    ">E5\n",
    "^ATCGTGGT\n",
    ">E7\n",
    "^CGGTAATC\n",
    ">E9\n",
    "^AGTTGTGC\n",
    ">E11\n",
    "^AATGACGC\n",
    ">E13\n",
    "^TACCGGAT\n",
    ">E15\n",
    "^TTGCAACG\n",
    ">E17\n",
    "^CACTTCAC\n",
    ">E19\n",
    "^TAGCCATG\n",
    ">E21\n",
    "^ACAGGCAT\n",
    ">E23\n",
    "^AGGTGTTG\n",
    ">G1\n",
    "^CAGTCACA\n",
    ">G3\n",
    "^TCGATGAC\n",
    ">G5\n",
    "^GAAGTGCT\n",
    ">G7\n",
    "^CTTCCTTC\n",
    ">G9\n",
    "^CGAACAAC\n",
    ">G11\n",
    "^AACAACCG\n",
    ">G13\n",
    "^ACCTCAGT\n",
    ">G15\n",
    "^CGTCTTCA\n",
    ">G17\n",
    "^TGCGTAAC\n",
    ">G19\n",
    "^AACACGCT\n",
    ">G21\n",
    "^ACTCGATC\n",
    ">G23\n",
    "^TGAGCTGT\n",
    ">I1\n",
    "^TACTGCTC\n",
    ">I3\n",
    "^GACGAACT\n",
    ">I5\n",
    "^CTTCGCAA\n",
    ">I7\n",
    "^ATGGCGAT\n",
    ">I9\n",
    "^ACATGCCA\n",
    ">I11\n",
    "^GTCAACAG\n",
    ">I13\n",
    "^GTGGTATG\n",
    ">I15\n",
    "^CCAACTTC\n",
    ">I17\n",
    "^GACGTCAT\n",
    ">I19\n",
    "^ACGTCCAA\n",
    ">I21\n",
    "^GATCCACT\n",
    ">I23\n",
    "^AGCCTATC\n",
    ">K1\n",
    "^AGCTACCA\n",
    ">K3\n",
    "^AGATTGCG\n",
    ">K5\n",
    "^CACACATC\n",
    ">K7\n",
    "^GAGCAATC\n",
    ">K9\n",
    "^ATAGAGCG\n",
    ">K11\n",
    "^GACCGATA\n",
    ">K13\n",
    "^CAGACGTT\n",
    ">K15\n",
    "^CTGAACGT\n",
    ">K17\n",
    "^TTGGACTG\n",
    ">K19\n",
    "^GTCTGCAA\n",
    ">K21\n",
    "^CCACATTG\n",
    ">K23\n",
    "^GATGGAGT\n",
    ">M1\n",
    "^AGGTCAAC\n",
    ">M3\n",
    "^TACACACG\n",
    ">M5\n",
    "^CAAGTCGT\n",
    ">M7\n",
    "^AGCTAGTG\n",
    ">M9\n",
    "^CTCCTAGT\n",
    ">M11\n",
    "^ACTCCTAC\n",
    ">M13\n",
    "^CAATCAGG\n",
    ">M15\n",
    "^TCGTGCAT\n",
    ">M17\n",
    "^TAACGTCG\n",
    ">M19\n",
    "^AAGGCGTA\n",
    ">M21\n",
    "^TCTTACGG\n",
    ">M23\n",
    "^CGTGTGAT\n",
    ">O1\n",
    "^AACAGGTG\n",
    ">O3\n",
    "^AGTCGAAG\n",
    ">O5\n",
    "^TGGAAGCA\n",
    ">O7\n",
    "^CTCGTTCT\n",
    ">O9\n",
    "^ACGAGAAC\n",
    ">O11\n",
    "^AAGCCTGA\n",
    ">O13\n",
    "^CTACAAGG\n",
    ">O15\n",
    "^CGATGTTC\n",
    ">O17\n",
    "^ACCGGTTA\n",
    ">O19\n",
    "^GAACGGTT\n",
    ">O21\n",
    "^CTGTACCA\n",
    ">O23\n",
    "^GCGCATAT\n",
    ">A2\n",
    "^TGATAGGC\n",
    ">A4\n",
    "^CATCCAAG\n",
    ">A6\n",
    "^GTGAGACT\n",
    ">A8\n",
    "^CTGATGAG\n",
    ">A10\n",
    "^ACGGTACA\n",
    ">A12\n",
    "^CTCGACTT\n",
    ">A14\n",
    "^ACAACGTG\n",
    ">A16\n",
    "^TGCTGTGA\n",
    ">A18\n",
    "^CCAAGTAG\n",
    ">A20\n",
    "^AACTGAGG\n",
    ">A22\n",
    "^AGGTAGGA\n",
    ">A24\n",
    "^TTCGCCAT\n",
    ">C2\n",
    "^CAGGTAAG\n",
    ">C4\n",
    "^GTATCGAG\n",
    ">C6\n",
    "^TTCACGGA\n",
    ">C8\n",
    "^GAGCTCTA\n",
    ">C10\n",
    "^GTCAGTCA\n",
    ">C12\n",
    "^CACGTCTA\n",
    ">C14\n",
    "^AATTCCGG\n",
    ">C16\n",
    "^TCTAGGAG\n",
    ">C18\n",
    "^ATCCGTTG\n",
    ">C20\n",
    "^GATAGCCA\n",
    ">C22\n",
    "^TATGACCG\n",
    ">C24\n",
    "^CGATTGGA\n",
    ">E2\n",
    "^ACAAGCTC\n",
    ">E4\n",
    "^GAACCTTC\n",
    ">E6\n",
    "^AGCGAGAT\n",
    ">E8\n",
    "^CCGTAACT\n",
    ">E10\n",
    "^TCAGACAC\n",
    ">E12\n",
    "^CGAAGTCA\n",
    ">E14\n",
    "^GTGATCCA\n",
    ">E16\n",
    "^ACTGGTGT\n",
    ">E18\n",
    "^CTAACCTG\n",
    ">E20\n",
    "^AGCCAACT\n",
    ">E22\n",
    "^CCAGTTGA\n",
    ">E24\n",
    "^AAGTGCAG\n",
    ">G2\n",
    "^AACCGTGT\n",
    ">G4\n",
    "^CGCGTATT\n",
    ">G6\n",
    "^AGTTCGCA\n",
    ">G8\n",
    "^TAGTCAGC\n",
    ">G10\n",
    "^AACACCAC\n",
    ">G12\n",
    "^GTAAGCAC\n",
    ">G14\n",
    "^GTCCTTGA\n",
    ">G16\n",
    "^CAGGTTCA\n",
    ">G18\n",
    "^CCAACACT\n",
    ">G20\n",
    "^GAGAGTAC\n",
    ">G22\n",
    "^AGATACGG\n",
    ">G24\n",
    "^GTTCTTCG\n",
    ">I2\n",
    "^ATTCCGCT\n",
    ">I4\n",
    "^AAGCTCAC\n",
    ">I6\n",
    "^TGATCACG\n",
    ">I8\n",
    "^CAATGCGA\n",
    ">I10\n",
    "^ATGCGTCA\n",
    ">I12\n",
    "^TACATCGG\n",
    ">I14\n",
    "^ACTGCGAA\n",
    ">I16\n",
    "^TCTGTCGT\n",
    ">I18\n",
    "^CTCAAGCT\n",
    ">I20\n",
    "^AACCACTC\n",
    ">I22\n",
    "^CTTACAGC\n",
    ">I24\n",
    "^AGTCTTGG\n",
    ">K2\n",
    "^CACGCAAT\n",
    ">K4\n",
    "^AGCTTCAG\n",
    ">K6\n",
    "^CCTCGTTA\n",
    ">K8\n",
    "^TGAGACGA\n",
    ">K10\n",
    "^CACAGGAA\n",
    ">K12\n",
    "^ACTCAACG\n",
    ">K14\n",
    "^AAGCGACT\n",
    ">K16\n",
    "^CCTACCTA\n",
    ">K18\n",
    "^ATCTCCTG\n",
    ">K20\n",
    "^TCACGATG\n",
    ">K22\n",
    "^CCACAACA\n",
    ">K24\n",
    "^AGGTCTGT\n",
    ">M2\n",
    "^AGAAGGAC\n",
    ">M4\n",
    "^GCGTATCA\n",
    ">M6\n",
    "^CAACACAG\n",
    ">M8\n",
    "^TCCACGTT\n",
    ">M10\n",
    "^ATCGCAAC\n",
    ">M12\n",
    "^ACGTCGTT\n",
    ">M14\n",
    "^CGAATACG\n",
    ">M16\n",
    "^TGCTTGCT\n",
    ">M18\n",
    "^CTCGAACA\n",
    ">M20\n",
    "^ACATGGAG\n",
    ">M22\n",
    "^ACAAGACG\n",
    ">M24\n",
    "^CGCCTTAT\n",
    ">O2\n",
    "^AGCAGACA\n",
    ">O4\n",
    "^GTTAAGCG\n",
    ">O6\n",
    "^CATGGATC\n",
    ">O8\n",
    "^ACAGAGGT\n",
    ">O10\n",
    "^TAAGTGGC\n",
    ">O12\n",
    "^AGTCAGGT\n",
    ">O14\n",
    "^GCCTTAAC\n",
    ">O16\n",
    "^GTTGGCAT\n",
    ">O18\n",
    "^CAACCTCT\n",
    ">O20\n",
    "^TGGATGGT\n",
    ">O22\n",
    "^CTATCCAC\n",
    ">O24\n",
    "^GATCTCAG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A02a_cellbarcodes_subset2.fa\n",
    ">B1\n",
    "^GAACGAAG\n",
    ">B3\n",
    "^ACCTAGAC\n",
    ">B5\n",
    "^TACGACGT\n",
    ">B7\n",
    "^TTGAGCTC\n",
    ">B9\n",
    "^AGTACACG\n",
    ">B11\n",
    "^TGTCAGTG\n",
    ">B13\n",
    "^GACTACGA\n",
    ">B15\n",
    "^TTACGTGC\n",
    ">B17\n",
    "^ACTGCTTG\n",
    ">B19\n",
    "^GCCTATGT\n",
    ">B21\n",
    "^GTACCACA\n",
    ">B23\n",
    "^TAGTGGTG\n",
    ">D1\n",
    "^ATACGCAG\n",
    ">D3\n",
    "^AAGACCGT\n",
    ">D5\n",
    "^CTCCAATC\n",
    ">D7\n",
    "^TCTGGACA\n",
    ">D9\n",
    "^AACACTGG\n",
    ">D11\n",
    "^TTGGTGCA\n",
    ">D13\n",
    "^CCTGTCAA\n",
    ">D15\n",
    "^CTATGCCT\n",
    ">D17\n",
    "^TTCGGCTA\n",
    ">D19\n",
    "^ACCGACAA\n",
    ">D21\n",
    "^CGTAGATG\n",
    ">D23\n",
    "^CTGTATGC\n",
    ">F1\n",
    "^GTTGCTGT\n",
    ">F3\n",
    "^AGAACCAG\n",
    ">F5\n",
    "^GATGTCGA\n",
    ">F7\n",
    "^AGGAGGTT\n",
    ">F9\n",
    "^AATCGCTG\n",
    ">F11\n",
    "^AGTGACCT\n",
    ">F13\n",
    "^CGAATTGC\n",
    ">F15\n",
    "^CAAGAAGC\n",
    ">F17\n",
    "^CACCAGTT\n",
    ">F19\n",
    "^GTATTCCG\n",
    ">F21\n",
    "^TTCGAAGC\n",
    ">F23\n",
    "^AGACCTTG\n",
    ">H1\n",
    "^CCAAGGTT\n",
    ">H3\n",
    "^ACGTATGG\n",
    ">H5\n",
    "^AAGGACCA\n",
    ">H7\n",
    "^TATGCGGT\n",
    ">H9\n",
    "^AAGGAAGG\n",
    ">H11\n",
    "^AGCGTGTA\n",
    ">H13\n",
    "^TCTACGCA\n",
    ">H15\n",
    "^TGGCTCTT\n",
    ">H17\n",
    "^CCTTCCAT\n",
    ">H19\n",
    "^ATACTGGC\n",
    ">H21\n",
    "^AACCTACG\n",
    ">H23\n",
    "^CATACTCG\n",
    ">J1\n",
    "^TGCACTTG\n",
    ">J3\n",
    "^TCACTCGA\n",
    ">J5\n",
    "^CACTGTAG\n",
    ">J7\n",
    "^GTACGATC\n",
    ">J9\n",
    "^TGGTGAAG\n",
    ">J11\n",
    "^TAGCTGAG\n",
    ">J13\n",
    "^AGAGCAGA\n",
    ">J15\n",
    "^CTTCGGTT\n",
    ">J17\n",
    "^ACAACAGC\n",
    ">J19\n",
    "^AGCCGTAA\n",
    ">J21\n",
    "^CTCTTGTC\n",
    ">J23\n",
    "^CAGATCCT\n",
    ">L1\n",
    "^GATGCTAC\n",
    ">L3\n",
    "^AGGAACAC\n",
    ">L5\n",
    "^ACCATCCT\n",
    ">L7\n",
    "^GAACGTGA\n",
    ">L9\n",
    "^TAGAACGC\n",
    ">L11\n",
    "^AACCAGAG\n",
    ">L13\n",
    "^CGACCTAA\n",
    ">L15\n",
    "^CTCTCAGA\n",
    ">L17\n",
    "^AGGCTGAA\n",
    ">L19\n",
    "^ATCGGAGA\n",
    ">L21\n",
    "^GATACCTG\n",
    ">L23\n",
    "^TCCTGACT\n",
    ">N1\n",
    "^TCAGCCTT\n",
    ">N3\n",
    "^AAGCATCG\n",
    ">N5\n",
    "^GCCAATAC\n",
    ">N7\n",
    "^GACACAGT\n",
    ">N9\n",
    "^AAGAGGCA\n",
    ">N11\n",
    "^GAAGACTG\n",
    ">N13\n",
    "^CCGTTATG\n",
    ">N15\n",
    "^CTAGCAGT\n",
    ">N17\n",
    "^GCCAGAAT\n",
    ">N19\n",
    "^CGAGAGAA\n",
    ">N21\n",
    "^AACTCGGA\n",
    ">N23\n",
    "^ACAGTTCG\n",
    ">P1\n",
    "^TGACCGTT\n",
    ">P3\n",
    "^CATCTGCT\n",
    ">P5\n",
    "^CGCTGATA\n",
    ">P7\n",
    "^TCGTCTGA\n",
    ">P9\n",
    "^CACATGGT\n",
    ">P11\n",
    "^CGAGTTAG\n",
    ">P13\n",
    "^AGCTAAGC\n",
    ">P15\n",
    "^GTTCCATG\n",
    ">P17\n",
    "^GCATCCTA\n",
    ">P19\n",
    "^CCATGAAC\n",
    ">P21\n",
    "^ATCCACGA\n",
    ">P23\n",
    "^GAGAAGGT\n",
    ">B2\n",
    "^AGGCAATG\n",
    ">B4\n",
    "^TCACCTAG\n",
    ">B6\n",
    "^CATACGGA\n",
    ">B8\n",
    "^GTCATCGT\n",
    ">B10\n",
    "^TTACCGAC\n",
    ">B12\n",
    "^ACCTTCGA\n",
    ">B14\n",
    "^ACGCTTCT\n",
    ">B16\n",
    "^GAGTAGAG\n",
    ">B18\n",
    "^ATGCCTAG\n",
    ">B20\n",
    "^CAACTCCA\n",
    ">B22\n",
    "^AAGTCCTC\n",
    ">B24\n",
    "^GTCGATTG\n",
    ">D2\n",
    "^GCGTTAGA\n",
    ">D4\n",
    "^TTGCGAGA\n",
    ">D6\n",
    "^ACACCGAT\n",
    ">D8\n",
    "^CGTATCTC\n",
    ">D10\n",
    "^AAGGAGAC\n",
    ">D12\n",
    "^TGTCGACT\n",
    ">D14\n",
    "^TCAATCCG\n",
    ">D16\n",
    "^GACTTGTG\n",
    ">D18\n",
    "^CCGATGTA\n",
    ">D20\n",
    "^TAGGAGCT\n",
    ">D22\n",
    "^CAACGAGT\n",
    ">D24\n",
    "^TGTGTCAG\n",
    ">F2\n",
    "^CTAGGTTG\n",
    ">F4\n",
    "^GTGTCCTT\n",
    ">F6\n",
    "^TACCTGCA\n",
    ">F8\n",
    "^CCTTAGGT\n",
    ">F10\n",
    "^CACAGACT\n",
    ">F12\n",
    "^TCGAACCT\n",
    ">F14\n",
    "^GCATAGTC\n",
    ">F16\n",
    "^CTCCTGAA\n",
    ">F18\n",
    "^AACGCACA\n",
    ">F20\n",
    "^TAGTCTCG\n",
    ">F22\n",
    "^ACTCTGAG\n",
    ">F24\n",
    "^GTTATGGC\n",
    ">H2\n",
    "^CTCGGTAA\n",
    ">H4\n",
    "^TACAGAGC\n",
    ">H6\n",
    "^GCATAACG\n",
    ">H8\n",
    "^GATCAGAC\n",
    ">H10\n",
    "^CGCAACTA\n",
    ">H12\n",
    "^TCCGATCA\n",
    ">H14\n",
    "^CAACTTGG\n",
    ">H16\n",
    "^TCAGTAGG\n",
    ">H18\n",
    "^ACAGCAAG\n",
    ">H20\n",
    "^GAATGGCA\n",
    ">H22\n",
    "^CGGATCAA\n",
    ">H24\n",
    "^ACTGCACT\n",
    ">J2\n",
    "^CCTAAGTC\n",
    ">J4\n",
    "^TTCGTACG\n",
    ">J6\n",
    "^TCCTGGTA\n",
    ">J8\n",
    "^CATTGACG\n",
    ">J10\n",
    "^ACCTCTTC\n",
    ">J12\n",
    "^CATTCGTC\n",
    ">J14\n",
    "^TTCCTCCT\n",
    ">J16\n",
    "^GCTGTAAG\n",
    ">J18\n",
    "^GACATCTC\n",
    ">J20\n",
    "^CAACCGTA\n",
    ">J22\n",
    "^TGCGATAG\n",
    ">J24\n",
    "^TGGTTCGA\n",
    ">L2\n",
    "^AAGCGTTC\n",
    ">L4\n",
    "^CGATTCTG\n",
    ">L6\n",
    "^GCAACCAT\n",
    ">L8\n",
    "^AATCCAGC\n",
    ">L10\n",
    "^AGTGCATC\n",
    ">L12\n",
    "^GCATTGGT\n",
    ">L14\n",
    "^CTTAGGAC\n",
    ">L16\n",
    "^ATAGTCGG\n",
    ">L18\n",
    "^GAGACCAA\n",
    ">L20\n",
    "^AACAAGGC\n",
    ">L22\n",
    "^CCAGTATC\n",
    ">L24\n",
    "^CCTCGAAT\n",
    ">N2\n",
    "^CAACTGAC\n",
    ">N4\n",
    "^TGCTCTAC\n",
    ">N6\n",
    "^CATCACGT\n",
    ">N8\n",
    "^GCCACTTA\n",
    ">N10\n",
    "^GCTTCACA\n",
    ">N12\n",
    "^ACCGAATG\n",
    ">N14\n",
    "^CTCACCAA\n",
    ">N16\n",
    "^CAGAACTG\n",
    ">N18\n",
    "^AGAAGCCT\n",
    ">N20\n",
    "^CACGATTC\n",
    ">N22\n",
    "^AAGCTGGT\n",
    ">N24\n",
    "^GCAATGAG\n",
    ">P2\n",
    "^CTCTATCG\n",
    ">P4\n",
    "^ACTCTCCA\n",
    ">P6\n",
    "^CAGCATAC\n",
    ">P8\n",
    "^TACTCCAG\n",
    ">P10\n",
    "^GAGGCATT\n",
    ">P12\n",
    "^ACACCTCA\n",
    ">P14\n",
    "^CGCAATGT\n",
    ">P16\n",
    "^CCTAGAGA\n",
    ">P18\n",
    "^TACTAGCG\n",
    ">P20\n",
    "^CGTCCATT\n",
    ">P22\n",
    "^TCGCTATC\n",
    ">P24\n",
    "^AATGGTCG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A02a_demultiplex_fastq.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A02a_demultiplex.$JOB_ID.$TASK_ID\n",
    "#$ -j y\n",
    "#$ -l h_rt=6:00:00,h_data=24G\n",
    "#$ -t 1-16\n",
    "#$ -N A02a_demultip\n",
    "#$ -hold_jid_ad A01a_merge_lanes\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init -------------------------------------------------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snmCTseq # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths -----------------------------------------------------\n",
    "\n",
    "platename=$(awk -v linenum=$SGE_TASK_ID -F ',' '$1 == linenum' $metadat_plate | cut -d , -f 2)\n",
    "\n",
    "r1file=\"fastq_raw/${platename}_R1.fastq.gz\"\n",
    "r2file=\"fastq_raw/${platename}_R2.fastq.gz\"\n",
    "\n",
    "\n",
    "\n",
    "# apply perl script for each of two index files --------------------------------\n",
    "\n",
    "for index_file in {1..2}\n",
    "do\n",
    "    echo -e \"\\n\\ndemultiplexing ${r1file} and ${r2file} with index set ${index_file} of 2...\"\n",
    "    perl Scripts/A02a_demultiplex_fastq.pl $r1file $r2file \\\n",
    "    Scripts/A02a_cellbarcodes_subset${index_file}.fa\n",
    "done\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A02a_demultip' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo \" \"\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `date `\n",
    "echo \" \"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (A02b) check demultiplexing output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A02b_check_demultip.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A02b_check_demult.$JOB_ID\n",
    "#$ -j y\n",
    "#$ -l h_rt=1:00:00,h_data=4G\n",
    "#$ -N A02b_check_demult\n",
    "#$ -hold_jid A02a_demultip\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init -------------------------------------------------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snmCTseq # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths------------------------------------------------------\n",
    "\n",
    "query_metadat () {\n",
    "  awk -F',' -v targetcol=\"$1\" \\\n",
    "      'NR==1 {\n",
    "                for (i=1;i<=NF;i++) {\n",
    "                    if ($i==targetcol) {assayout=i; break} }\n",
    "                print $assayout\n",
    "              } \n",
    "      NR>1 {\n",
    "                print $assayout\n",
    "            }' $metadat_well\n",
    "}\n",
    "\n",
    "check_filepaths_in_assay() {\n",
    "    for file in $@\n",
    "        do \n",
    "        if [[ ! -s $file ]]\n",
    "            then\n",
    "                echo \"missing '$file'\"\n",
    "            fi\n",
    "        done\n",
    "}\n",
    "\n",
    "r1demultip=($(query_metadat \"A02a_fqgz_demultip_R1\"))\n",
    "r2demultip=($(query_metadat \"A02a_fqgz_demultip_R2\"))\n",
    "\n",
    "\n",
    "\n",
    "# apply checks for A01a output -------------------------------------------------\n",
    "# (may be later depreciated due to explicit target filename change)\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n-----------------------------------------------------------------\"\n",
    "echo \"A. counting number of .fastq.gz files generated during demultiplexing step... \"\n",
    "echo \"-----------------------------------------------------------------\"\n",
    "\n",
    "nfastq_tot_demultip=0\n",
    "\n",
    "echo -e \"\\nR1.fastq.gz files\\n\"\n",
    "for prefix in $(cut -f 2 -d ',' $metadat_plate | tail -n +2);\n",
    "    do\n",
    "        nfastq_plate=$(\n",
    "            find \"fastq_demultip/\" -maxdepth 1 -mindepth 1 -name ${prefix}*R1.fastq.gz | wc -l)\n",
    "        echo \"$prefix : $nfastq_plate\"\n",
    "        let \"nfastq_tot_demultip+=$nfastq_plate\"\n",
    "done\n",
    "\n",
    "echo -e \"\\nR2.fastq.gz files\\n\"\n",
    "for prefix in $(cut -f 2 -d ',' $metadat_plate | tail -n +2);\n",
    "    do\n",
    "        nfastq_plate=$(\n",
    "            find \"fastq_demultip/\" -maxdepth 1 -mindepth 1 -name ${prefix}*R2.fastq.gz | wc -l)\n",
    "        echo \"$prefix : $nfastq_plate\"\n",
    "        let \"nfastq_tot_demultip+=$nfastq_plate\"\n",
    "done\n",
    "\n",
    "num_platesin=$(cut -f 2 -d ',' $metadat_plate | tail -n +2 | wc -l)\n",
    "nfastq_expected=$(($num_platesin*384*2))\n",
    "nfastq_demult_folder=$(find \"fastq_demultip/\" -maxdepth 1 -mindepth 1 -name '*.fastq.gz' | wc -l)\n",
    "fastq_missing=$(($nfastq_expected - $nfastq_tot_demultip))\n",
    "\n",
    "echo -e \"* num demultiplexed .fastqs expected: ($num_platesin plates)*384*2 = $nfastq_expected R1 & R2\"\n",
    "echo -e \"\\n* num detected, summed from per-plate values above: ${nfastq_tot_demultip}\"\n",
    "echo \"* num .fastqs in fastq_demultip: ${nfastq_demult_folder} (regardless of expected plate prefix; should be same as above)\"\n",
    "echo \"* num .fastqs missing: ${fastq_missing} ($((${fastq_missing}/2)) empty wells?)\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n-----------------------------------------------------------------\"\n",
    "echo \"B. tabulating number of empty wells detected by perl script per plate... \"\n",
    "echo -e \"-----------------------------------------------------------------\\n\"\n",
    "\n",
    "grep \"rm fastq\" -c sublogs/A02a*\n",
    "\n",
    "echo -e \"\\n* # missing wells, based on # demultiplexing indices undetected in perl script.\"\n",
    "echo \"* should be consistent with [A], with same number of wells as detected cell barcodes.\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n-----------------------------------------------------------------\"\n",
    "echo \"C. checking expected output from well-level metadata file (A01c)...\"\n",
    "echo \"-----------------------------------------------------------------\"\n",
    "\n",
    "echo -e \"\\nchecking Read 1 .fastq.gz files:\\n\"\n",
    "check_filepaths_in_assay ${r1demultip[@]}\n",
    "\n",
    "echo -e \"\\nchecking Read 2 .fastq.gz files:\\n\"\n",
    "check_filepaths_in_assay ${r2demultip[@]}\n",
    "\n",
    "echo -e \"\\n* checks the A02a output columns of 'metadat_well' if the file exists and is non-empty.\"\n",
    "echo \"* if none missing, will only output target column names above.\"\n",
    "echo \"* if some declared 'missing' but all other checks OK, cell barcode may not exist.\"\n",
    "echo \"  (check 'fastq_demultip/plate_summary.txt report)\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n-----------------------------------------------------------------\"\n",
    "echo \"D. checking that each .fastq in demultiplex folder has a paired-end (R1/R2) mate...\"\n",
    "echo \"-----------------------------------------------------------------\"\n",
    "\n",
    "nfastq_no_pairmate=0\n",
    "for r1file in `find \"fastq_demultip/\" -maxdepth 1 -mindepth 1 -name '*indexed_R1.fastq.gz'`;\n",
    "do\n",
    "        if [ ! -e \"${r1file/R1/R2}\" ]\n",
    "            then\n",
    "                echo \"missing after demultiplexing: ${r1file/R1/R2}\"\n",
    "                let \"nfastq_no_pairmate+=1\"\n",
    "            fi\n",
    "        done\n",
    "for r2file in `find \"fastq_demultip/\" -maxdepth 1 -mindepth 1 -name '*indexed_R2.fastq.gz'`;\n",
    "do\n",
    "        if [ ! -e \"${r2file/R2/R1}\" ]\n",
    "            then\n",
    "                echo \"missing after demultiplexing: ${r2file/R2/R1}\"\n",
    "                let \"nfastq_no_pairmate+=1\"\n",
    "            fi\n",
    "        done\n",
    "\n",
    "echo -e \"\\n* $nfastq_no_pairmate cases where a well has their paired-end mate missing after demultiplexing.\"\n",
    "echo \"* if non-zero, there are wells with their R1 but no corresponding R2 reads, or their R2 but no R1.\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n-----------------------------------------------------------------\"\n",
    "echo \"E. checking that two summary .txt files were generated for each file... \"\n",
    "echo -e \"-----------------------------------------------------------------\\n\"\n",
    "\n",
    "summary_txt_missing=0\n",
    "for prefix in $(cut -f 2 -d ',' $metadat_plate | tail -n +2);\n",
    "    do\n",
    "        if [[ $(find \"fastq_demultip/\" -maxdepth 1 -name \"${prefix}*.txt\" | wc -l) != 2 ]]\n",
    "            then \n",
    "            let \"summary_txt_missing+=1\"\n",
    "            echo \"summary files missing for : ${prefix}\"\n",
    "        fi\n",
    "    done\n",
    "\n",
    "echo -e \"* $summary_txt_missing 'raw_fastq' files are missing their summary .txt file(s).\"\n",
    "echo \"* if any summaries are missing, the demultiplexing jobs may have failed (historically, memory issues)\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n-----------------------------------------------------------------\"\n",
    "echo \"F. printing percent barcodes unassigned...\"\n",
    "echo -e \"-----------------------------------------------------------------\\n\"\n",
    "\n",
    "for prefix in $(cut -f 2 -d ',' $metadat_plate | tail -n +2);\n",
    "do\n",
    "    echo $prefix\n",
    "    for summaryfile in fastq_demultip/${prefix}*summary*.txt\n",
    "    do\n",
    "        awk 'NR==2' $summaryfile | cut -f 4\n",
    "    done\n",
    "done\n",
    "\n",
    "echo -e \"\\n* unassigned reads should be ~50%, assuming two barcode ref files (summary_1.txt, summary_2.txt)\"\n",
    "echo \"  caveat: sum generally >100% b/c of reads 'double counted' as unassigned to both barcode lists.\"\n",
    "echo \"* total unassigned expected 1-3%, usually due to ambig bases (N) or sequencing errors;\"\n",
    "echo \"  higher rates suggestive of contaminating library, issues in lane merging (fastq_raw input)\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A02b_check_demultip' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "echo \" \"\n",
    "echo \"Job $JOB_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID ended on:   \" `date `\n",
    "echo \" \"\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (A02c) fastqc on random well subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A02c_fastqc_demultip_fastq.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A02c_fastqc_demultip.$JOB_ID\n",
    "#$ -j y\n",
    "#$ -N A02c_fastqc_demultip\n",
    "#$ -l h_rt=8:00:00,h_data=8G\n",
    "#$ -pe shared 4\n",
    "#$ -hold_jid A02a_demultip\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init ------------------—------------------—-----------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snmCTseq # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "\n",
    "# fastqc specific parameters\n",
    "n_wells=4  # <-- num wells/plate to sample\n",
    "overwrite_random_wells=false # <-- overwrite target list of wells (false by default)\n",
    "\n",
    "mkdir qc_demultip\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# choose random wells ------------------—------------------—--------------------\n",
    "\n",
    "# randomly select $n_wells from each prefix associated with 'raw_fastq' R1s to run fastqc on\n",
    "# & saves the names of the corresponding demultiplexed R1 .fastq in \"$wells_to_run\"\n",
    "\n",
    "# warning: will overwrite this .txt if already exists,\n",
    "# to avoid re-running more files than expected / the same files multiple times\n",
    "\n",
    "if [[ -e $wells_to_run && \"$overwrite_random_wells\" == true ]]\n",
    "    then\n",
    "    \n",
    "        echo \"warning: Scripts/A02c_random_fastqc_wells.txt already exists.\"\n",
    "        echo \"deleting and re-randomizing. (overwrite_random_wells=true)\"\n",
    "\n",
    "        rm $wells_to_run\n",
    "fi\n",
    "\n",
    "# make random samples (otherwise skip)\n",
    "if [[ ! -e $wells_to_run ]]\n",
    "then\n",
    "    for prefix in $(cut -f 2 -d ',' $metadat_plate | tail -n +2);\n",
    "        do\n",
    "        printf '%s\\n' $(ls fastq_demultip/${prefix}*R1*fastq.gz \\\n",
    "                        | shuf \\\n",
    "                        | head -n $n_wells) \\\n",
    "             >> $wells_to_run\n",
    "        done\n",
    "else\n",
    "    echo \"warning: Scripts/A02c_random_fastqc_wells.txt already exists.\"\n",
    "    echo \"using existing random wells. (overwrite_random_wells=false)\"\n",
    "fi\n",
    "\n",
    "# print wells for record in sublog\n",
    "echo -e \"\\nnum wells sampled: $(wc -l $wells_to_run)\\n\"\n",
    "cat $wells_to_run\n",
    "\n",
    "\n",
    "\n",
    "# apply fastqc ------------------—------------------—---------------------------\n",
    "\n",
    "# run fastqc on the randomly selected wells\n",
    "echo \"running fastqc on each well.\"\n",
    "while read r1file;\n",
    "    do\n",
    "        r2file=${r1file/R1/R2}\n",
    "        fastqc -t 4 ${r1file} -o qc_demultip/\n",
    "        fastqc -t 4 ${r2file} -o qc_demultip/\n",
    "    done < $wells_to_run\n",
    "\n",
    "# compile across wells\n",
    "echo \"aggregating via multiqc.\"\n",
    "multiqc -d qc_demultip -o qc_demultip -n multiqc_fastqc \\\n",
    "    --cl-config \"sp: { fastqc/zip: {fn: '*_fastqc.zip' } }\" -m \"fastqc\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A02c_fastqc_demultip_fastq' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "echo \" \"\n",
    "echo \"Job $JOB_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID ended on:   \" `date `\n",
    "echo \" \"\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
