{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## A03_trimming\n",
    "# ## overall commands\n",
    "\n",
    "# qsub Scripts/A03a_trimming_fastp.sub # †\n",
    "# qsub Scripts/A03b_check_trimmed.sub # ‡\n",
    "# qsub Scripts/A03c_fastqc_trimmed.sub\n",
    "\n",
    "\n",
    "# # * = job array based on \"platenum\"\n",
    "# # † = job array based on \"batchnum\" (two rows at a time)\n",
    "# # ‡ fast enough to run interactively"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (A03a) fastp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A03a_adapter_sequences.fa\n",
    ">P5_pre\n",
    "AATGATACGGCGACCACCGAGATCTACAC\n",
    ">P5_post\n",
    "ACTCTTTCCCTACACGACGCTCT\n",
    ">P5_pre_rc\n",
    "GTGTAGATCTCGGTGGTCGCCGTATCATT\n",
    ">P5_post_rc\n",
    "AGAGCGTCGTGTAGGGAAAGAGT\n",
    ">P7_pre\n",
    "CAAGCAGAAGACGGCATACGAGAT\n",
    ">P7_post\n",
    "ACTGGAGTTCAGACGTGTGCTCTT\n",
    ">P7_pre_rc\n",
    "ATCTCGTATGCCGTCTTCTGCTTG\n",
    ">P7_post_rc\n",
    "AAGAGCACACGTCTGAACTCCAGT\n",
    ">TruSeq_universal\n",
    "AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT\n",
    ">TruSeq_universal_rc\n",
    "AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATT\n",
    ">IlmnBisulfite_fwd\n",
    "AGATCGGAAGAGCACACGTCTGAAC\n",
    ">IlmnBisulfite_rev\n",
    "AGATCGGAAGAGCGTCGTGTAGGGA\n",
    ">IlmnBisulfite_fwd_rc\n",
    "GTTCAGACGTGTGCTCTTCCGATCT\n",
    ">IlmnBisulfite_rev_rc\n",
    "TCCCTACACGACGCTCTTCCGATCT\n",
    ">SmartSeq_fwd\n",
    "AAGCAGTGGTATCAACGCAGAGT\n",
    ">SmartSeq_rev\n",
    "ACTCTGCGTTGATACCACTGCTT\n",
    ">SmartSeq_fwd_rc\n",
    "ACTCTGCGTTGATACCACTGCTT\n",
    ">SmartSeq_rev_rc\n",
    "AAGCAGTGGTATCAACGCAGAGT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A03a_trimming_fastp.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A03a_trim.$JOB_ID.$TASK_ID\n",
    "#$ -j y\n",
    "#$ -l h_rt=8:00:00,h_data=8G\n",
    "#$ -N A03a_trim\n",
    "#$ -t 1-512\n",
    "#$ -hold_jid A02a_demultip\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init -------------------------------------------------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snmCTseq # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "overwrite_existing=false # <-- for help with incomplete jobs\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths -----------------------------------------------------\n",
    "\n",
    "# helper functions\n",
    "query_metadat () {\n",
    "  awk -F',' -v targetcol=\"$1\" \\\n",
    "  'NR==1 {\n",
    "                for (i=1;i<=NF;i++) {\n",
    "                    if ($i==targetcol) {assayout=i; break} }\n",
    "                print $assayout\n",
    "              }\n",
    "      NR>1 {\n",
    "                print $assayout\n",
    "            }' ${metadat_well}\n",
    "}\n",
    "\n",
    "# extract target wells, print values for log\n",
    "\n",
    "batchnum=($(query_metadat \"batchnum\"))\n",
    "nwells=${#batchnum[@]}\n",
    "  \n",
    "target_well_rows=()\n",
    "for ((row=1; row<=nwells; row++))\n",
    "do\n",
    "    if [[ \"${batchnum[$row]}\" == \"${SGE_TASK_ID}\" ]]\n",
    "    then\n",
    "        target_well_rows+=($row)\n",
    "    fi\n",
    "done\n",
    "\n",
    "# filepaths associated with target rows in well-level metadata\n",
    "\n",
    "wellprefix=($(query_metadat \"wellprefix\"))\n",
    "\n",
    "r1in=($(query_metadat \"A02a_fqgz_demultip_R1\"))\n",
    "r2in=($(query_metadat \"A02a_fqgz_demultip_R2\"))\n",
    "\n",
    "r1paired=($(query_metadat \"A03a_fqgz_paired_R1\"))\n",
    "r2paired=($(query_metadat \"A03a_fqgz_paired_R2\"))\n",
    "\n",
    "r1singletrim=($(query_metadat \"A03a_fqgz_singletrim_R1\"))\n",
    "r2singletrim=($(query_metadat \"A03a_fqgz_singletrim_R2\"))\n",
    "\n",
    "jsonout=($(query_metadat \"A03a_json_fastp\"))\n",
    "\n",
    "  \n",
    "  \n",
    "# run fastp for trimming -----------------------------------------------------\n",
    "\n",
    "echo \"batch number: ${SGE_TASK_ID}\"\n",
    "echo \"processing the following rows in well metadata file (${metadat_well}):\"\n",
    "\n",
    "for row in ${target_well_rows[@]}\n",
    "do\n",
    "    echo -e \"$row\\t${wellprefix[$row]}\"\n",
    "done\n",
    "\n",
    "if [[ ! -s fastq_trimmed ]]\n",
    "then\n",
    "    mkdir fastq_trimmed\n",
    "fi\n",
    "\n",
    "# trim each well in batch (could be empty wells w/o demultip output,\n",
    "# but since by well doesn't cause catastrophic problems)\n",
    "for row in ${target_well_rows[@]}\n",
    "do\n",
    "\n",
    "    echo -e \"\\n\\ntrimming '${wellprefix[$row]}'...\\n\\n\"\n",
    "\n",
    "    # check for existing trimming output\n",
    "    # if well already processed, skip; if file doesn't exist, run fastp\n",
    "    if [[ -s ${r1paired[$row]} && -s ${r2paired[$row]} &&\n",
    "          -s ${r1singletrim[$row]} && -s ${r2singletrim[$row]} && \\\n",
    "          -s ${jsonout[$row]} ]]\n",
    "    then\n",
    "        echo -e \"output files for '${wellprefix[$row]}' already exist.\"\n",
    "        if [[ \"$overwrite_existing\" == \"true\" ]]\n",
    "        then\n",
    "            echo \"overwrite_existing=true. trimming anyway.\"\n",
    "        else\n",
    "            echo \"overwrite_existing=false. skipping.\"\n",
    "            continue\n",
    "        fi\n",
    "    fi\n",
    "    \n",
    "    fastp -i ${r1in[$row]} -I ${r2in[$row]} \\\n",
    "        -o ${r1paired[$row]} -O ${r2paired[$row]} \\\n",
    "        --unpaired1 ${r1singletrim[$row]} --unpaired2 ${r2singletrim[$row]} \\\n",
    "        -h fastq_trimmed/${wellprefix[$row]}.html -j ${jsonout[$row]} -R ${wellprefix[$row]} \\\n",
    "        --adapter_fasta=Scripts/A03a_adapter_sequences.fa \\\n",
    "        -f 17 -t 10 -F 15 -T 10 -l 30 \\\n",
    "        --cut_right -q 20 -u 50 -y -Y 15 -x\n",
    "    \n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A03a_trim' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `date `\n",
    "echo \" \""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (A03b) file check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A03b_check_trimmed.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A03b_check_trimmed.$JOB_ID\n",
    "#$ -j y\n",
    "#$ -l h_rt=1:00:00,h_data=8G\n",
    "#$ -N A03b_check_trimmed\n",
    "#$ -hold_jid A03a_trim\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init -------------------------------------------------------------\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "overwrite_existing=false # <-- for help with incomplete jobs\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths -----------------------------------------------------\n",
    "\n",
    "query_metadat () {\n",
    "  awk -F',' -v targetcol=\"$1\" \\\n",
    "  'NR==1 {\n",
    "                for (i=1;i<=NF;i++) {\n",
    "                    if ($i==targetcol) {assayout=i; break} }\n",
    "                print $assayout\n",
    "              } \n",
    "      NR>1 {\n",
    "                print $assayout\n",
    "            }' ${metadat_well}\n",
    "}\n",
    "\n",
    "check_filepaths_in_assay() {\n",
    "    for file in $@\n",
    "    do \n",
    "        if [[ ! -s ${file} ]]\n",
    "        then\n",
    "            echo \"missing '${file}'\"\n",
    "        fi\n",
    "    done\n",
    "}\n",
    "\n",
    "check_filepath_by_batch() {\n",
    "target_array=($@)\n",
    "batches_to_rerun=()\n",
    "for ((target_batch=1; target_batch<=nbatches; target_batch++))\n",
    "    do\n",
    "        target_well_rows=()\n",
    "        for ((row=1; row<=nwells; row++))\n",
    "        do\n",
    "            if [[ \"${batchnum[$row]}\" == \"${target_batch}\" ]]\n",
    "            then\n",
    "                target_well_rows+=($row)\n",
    "            fi\n",
    "        done\n",
    "    \n",
    "        batch_file_list=${target_array[@]: ${target_well_rows[0]}:${#target_well_rows[@]} }\n",
    "    \n",
    "        num_files_missing=$(check_filepaths_in_assay ${batch_file_list[@]} | wc -l)\n",
    "\n",
    "        if [[ ${num_files_missing} > 0 ]]\n",
    "        then\n",
    "            batches_to_rerun+=(${target_batch})\n",
    "            echo -e \"${target_batch} \\t ${num_files_missing}\"\n",
    "        fi\n",
    "    done \n",
    "    \n",
    "    if [[ ${#batches_to_rerun[@]} > 0 ]]\n",
    "    then\n",
    "        echo \"batches to re-run:\"\n",
    "        echo \"${batches_to_rerun[*]}\"        \n",
    "    fi\n",
    "}\n",
    "\n",
    "batchnum=($(query_metadat \"batchnum\"))\n",
    "\n",
    "nwells=${#batchnum[@]}\n",
    "nbatches=${batchnum[-1]}\n",
    "\n",
    "r1paired=($(query_metadat \"A03a_fqgz_paired_R1\"))\n",
    "r2paired=($(query_metadat \"A03a_fqgz_paired_R2\"))\n",
    "\n",
    "r1singletrim=($(query_metadat \"A03a_fqgz_singletrim_R1\"))\n",
    "r2singletrim=($(query_metadat \"A03a_fqgz_singletrim_R2\"))\n",
    "\n",
    "jsonout=($(query_metadat \"A03a_json_fastp\"))\n",
    "\n",
    "\n",
    "\n",
    "# apply checks for A03a output -------------------------------------------------\n",
    "\n",
    "echo \"-----------------------------------------------------------------\"\n",
    "echo \"A. counting number of .fastq.gz files generated during trimming step... \"\n",
    "echo \"-----------------------------------------------------------------\"\n",
    "\n",
    "\n",
    "nfastq_tot_trimmed=0\n",
    "\n",
    "num_platesin=$(cut -f 2 -d ',' ${metadat_plate} | tail -n +2 | wc -l)\n",
    "nfastq_expected=$((${num_platesin}*384*4))\n",
    "platenames=($(cut -f 2 -d ',' ${metadat_plate} | tail -n +2))\n",
    "\n",
    "echo -e \"\\npaired_R1.fastq.gz files\\n\"\n",
    "for ((i=0; i<${num_platesin}; i++)); \n",
    "    do\n",
    "        nfastq_plate=$(\n",
    "            find \"fastq_trimmed/\" -maxdepth 1 -mindepth 1 -name ${platenames[$i]}*paired_R1.fastq.gz | wc -l)\n",
    "        echo -e \"$(($i + 1)) \\t ${platenames[$i]} \\t ${nfastq_plate}\"\n",
    "        let \"nfastq_tot_trimmed+=${nfastq_plate}\"\n",
    "done\n",
    "\n",
    "echo -e \"\\npaired_R2.fastq.gz files\\n\"\n",
    "for ((i=0; i<${num_platesin}; i++)); \n",
    "    do\n",
    "        nfastq_plate=$(\n",
    "            find \"fastq_trimmed/\" -maxdepth 1 -mindepth 1 -name ${platenames[$i]}*paired_R2.fastq.gz | wc -l)\n",
    "        echo -e \"$(($i + 1)) \\t ${platenames[$i]} \\t ${nfastq_plate}\"\n",
    "        let \"nfastq_tot_trimmed+=${nfastq_plate}\"\n",
    "done\n",
    "\n",
    "echo -e \"\\nsingletrim_R1.fastq.gz files\\n\"\n",
    "for ((i=0; i<${num_platesin}; i++)); \n",
    "    do\n",
    "        nfastq_plate=$(\n",
    "            find \"fastq_trimmed/\" -maxdepth 1 -mindepth 1 -name ${platenames[$i]}*singletrim_R1.fastq.gz | wc -l)\n",
    "        echo -e \"$(($i + 1)) \\t ${platenames[$i]} \\t ${nfastq_plate}\"\n",
    "        let \"nfastq_tot_trimmed+=${nfastq_plate}\"\n",
    "done\n",
    "\n",
    "echo -e \"\\nsingletrim_R2.fastq.gz files\\n\"\n",
    "for ((i=0; i<${num_platesin}; i++)); \n",
    "    do\n",
    "        nfastq_plate=$(\n",
    "            find \"fastq_trimmed/\" -maxdepth 1 -mindepth 1 -name ${platenames[$i]}*singletrim_R2.fastq.gz | wc -l)\n",
    "        echo -e \"$(($i + 1)) \\t ${platenames[$i]} \\t ${nfastq_plate}\"\n",
    "        let \"nfastq_tot_trimmed+=${nfastq_plate}\"\n",
    "done\n",
    "\n",
    "nfastq_demult_folder=$(find \"fastq_demultip/\" -maxdepth 1 -mindepth 1 -name '*.fastq.gz' | wc -l)\n",
    "nfastq_trimmed_folder=$(find \"fastq_trimmed/\" -maxdepth 1 -mindepth 1 -name '*.fastq.gz' | wc -l)\n",
    "fastq_missing=$((${nfastq_expected} - ${nfastq_tot_trimmed}))\n",
    "\n",
    "echo -e \"\\n\\nchecks based on plate metadata:\"\n",
    "echo \"* num trimmed .fastqs expected based on # plates: (${num_platesin} plates)*384*4 = ${nfastq_expected} (R1p, R2p, R1single, R2single)\"\n",
    "echo \"* num trimmed .fastq detected, summed from per-plate values above: ${nfastq_tot_trimmed}\"\n",
    "echo \"* num .fastqs missing: ${fastq_missing} ($((${fastq_missing}/4)) wells have no trimmed output?)\"\n",
    "\n",
    "echo -e \"\\nother checks:\"\n",
    "echo \"* num .fastqs in fastq_trimmed: ${nfastq_trimmed_folder} (regardless of expected plate prefix; should be same as above)\"\n",
    "echo \"* num demultiplexed .fastqs detected: ${nfastq_demult_folder} (R1 & R2, count*2 = $((${nfastq_demult_folder}*2)) should equal trimmed .fastq.gz counts)\"\n",
    "echo -e \"\\n\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "echo \"-----------------------------------------------------------------\"\n",
    "echo \"B. checking each expected file (from ${metadat_well}, by batch...)\"\n",
    "echo \"-----------------------------------------------------------------\"\n",
    "\n",
    "r1paired=($(query_metadat \"A03a_fqgz_paired_R1\"))\n",
    "r2paired=($(query_metadat \"A03a_fqgz_paired_R2\"))\n",
    "r1singletrim=($(query_metadat \"A03a_fqgz_singletrim_R1\"))\n",
    "r2singletrim=($(query_metadat \"A03a_fqgz_singletrim_R2\"))\n",
    "jsonout=($(query_metadat \"A03a_json_fastp\"))\n",
    "\n",
    "echo -e \"\\nchecking Read 1 properly paired .fastq.gz files:\\n\"\n",
    "echo -e \"batchnum\\tnum_missing\"\n",
    "check_filepath_by_batch ${r1paired[@]}\n",
    "\n",
    "echo -e \"\\nchecking Read 2 properly paired .fastq.gz files:\\n\"\n",
    "echo -e \"batchnum\\tnum_missing\"\n",
    "check_filepath_by_batch ${r2paired[@]}\n",
    "\n",
    "echo -e \"\\nchecking Read 1 trimming singleton .fastq.gz files:\\n\"\n",
    "echo -e \"batchnum\\tnum_missing\"\n",
    "check_filepath_by_batch ${r1singletrim[@]}\n",
    "\n",
    "echo -e \"\\nchecking Read 2 trimming singleton .fastq.gz files:\\n\"\n",
    "echo -e \"batchnum\\tnum_missing\"\n",
    "check_filepath_by_batch ${r2singletrim[@]}\n",
    "\n",
    "echo -e \"\\nchecking .json summary file out:\\n\"\n",
    "echo -e \"batchnum\\tnum_missing\"\n",
    "check_filepath_by_batch ${jsonout[@]}\n",
    "\n",
    "echo -e \"\\n\\nsuggest re-running and checking sublog output of above batches.\"\n",
    "\n",
    "\n",
    "echo -e \"\\n* checks the A03a output columns of 'metadat_well' if the file exists and is non-empty.\"\n",
    "echo \"* if none missing, will only output target column names above.\"\n",
    "echo \"* caveat: if some fastq.gz files seem 'missing' but all other checks OK, may just be no/few reads surviving trimming.\"\n",
    "echo \"  (check 'fastq_demultip/' and associated fastp logs e.g., fastq_trimmed/wellprefix.html report)\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n-----------------------------------------------------------------\"\n",
    "echo \"C. checking log files for issues.\"\n",
    "echo -e \"-----------------------------------------------------------------\\n\"\n",
    "\n",
    "echo \"checking if 'completed' in sublogs/A03a_trim* output.\"\n",
    "echo \"if any filename is printed, the associated batch may have not completed trimming.\"\n",
    "\n",
    "grep -c \"'A03a_trim' completed\" sublogs/A03a_trim* | awk -F \":\" '$2==0 {print $1}'\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A03b_check_trimmed' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "echo \" \"\n",
    "echo \"Job $JOB_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID ended on:   \" `date `\n",
    "echo \" \""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (A03c) fastqc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A03c_fastqc_trimmed.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A03c_fastqc_trimmed.$JOB_ID\n",
    "#$ -j y\n",
    "#$ -N A03c_fastqc_trimmed\n",
    "#$ -l h_rt=8:00:00,h_data=8G\n",
    "#$ -pe shared 4\n",
    "#$ -hold_jid A03a_trim,A02c_fastqc_demultip\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init -------------------------------------------------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snmCTseq # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "overwrite_existing=false # <-- for help with incomplete jobs\n",
    "wells_to_run=Scripts/A02c_random_fastqc_wells.txt # <-- (qc'd wells from A02c by default)\n",
    "n_wells=4  # <-- num wells/plate to sample\n",
    "overwrite_random_wells=false # <-- overwrite target list of wells (false by default)\n",
    "\n",
    "\n",
    "\n",
    "# choose random wells ----------------------------------------------------------\n",
    "\n",
    "# randomly select $n_wells from each prefix associated with 'raw_fastq' R1s to run fastqc on\n",
    "# & saves the names of the corresponding R1 .fastq in \"${wells_to_run}\"\n",
    "\n",
    "# warning: will overwrite this .txt if already exists,\n",
    "# to avoid re-running more files than expected / the same files multiple times\n",
    "\n",
    "if [[ ! -s qc_trimmed ]]\n",
    "then\n",
    "    mkdir qc_trimmed\n",
    "fi\n",
    "\n",
    "if [[ -e ${wells_to_run} && \"${overwrite_random_wells}\" == true ]]\n",
    "    then\n",
    "    \n",
    "        echo \"warning: Scripts/A02c_random_fastqc_wells.txt already exists.\"\n",
    "        echo \"deleting and re-randomizing. (overwrite_random_wells=true)\"\n",
    "\n",
    "        rm ${wells_to_run}\n",
    "fi\n",
    "\n",
    "# make random samples (otherwise skip)\n",
    "if [[ ! -e ${wells_to_run} ]]\n",
    "then\n",
    "    for prefix in $(cut -f 2 -d ',' ${metadat_plate} | tail -n +2);\n",
    "        do\n",
    "        printf '%s\\n' $(ls fastq_trimmed/${prefix}*R1*fastq.gz \\\n",
    "                        | shuf \\\n",
    "                        | head -n ${n_wells}) \\\n",
    "             >> ${wells_to_run}\n",
    "        done\n",
    "else\n",
    "    echo \"warning: Scripts/A02c_random_fastqc_wells.txt already exists.\"\n",
    "    echo \"using existing random wells. (overwrite_random_wells=false)\"\n",
    "fi\n",
    "\n",
    "# print wells for record in sublog\n",
    "echo -e \"\\nnum wells sampled: $(wc -l ${wells_to_run})\\n\"\n",
    "cat ${wells_to_run}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# apply fastqc -----------------------------------------------------------------\n",
    "\n",
    "# run fastqc on the randomly selected wells\n",
    "echo \"running fastqc on each well.\"\n",
    "while read r1file;\n",
    "    do\n",
    "        wellprefix=$(echo $(basename ${r1file}) | awk -F\"_indexed\" '{print $1}')\n",
    "        echo $(ls fastq_trimmed/${wellprefix})\n",
    "        fastqc -t 4 fastq_trimmed/${wellprefix}*fastq.gz -o qc_trimmed/\n",
    "    done < ${wells_to_run}\n",
    "\n",
    "# compile across wells\n",
    "echo \"aggregating via multiqc.\"\n",
    "multiqc -d qc_trimmed -o qc_trimmed -n multiqc_fastqc \\\n",
    "    --cl-config \"sp: { fastqc/zip: {fn: '*_fastqc.zip' } }\" -m \"fastqc\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A03c_fastqc_trimmed' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "echo \" \"\n",
    "echo \"Job $JOB_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID ended on:   \" `date `\n",
    "echo \" \""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
