{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## overall commands\n",
    "# ## might be cleaner to wait for bismark mapping to finish 100% (run A04a only --> check output)\n",
    "# ## in lieu of submitting all of the below, but technically could run all cmds at once\n",
    "\n",
    "\n",
    "# qsub Scripts/A04a_bismark_map_to_allc.sub # †\n",
    "# qsub Scripts/A04b_check_bismark.sub \n",
    "# qsub Scripts/A04c_allc_to_mcds.sub  # *\n",
    "# qsub Scripts/A04d_calc_mC_fracs.sub  # †\n",
    "# qsub Scripts/A04e_samstats_DNA.sub  # †\n",
    "# qsub Scripts/A04f_coverage_DNA.sub # †\n",
    "\n",
    "\n",
    "# # * = job array based on \"platenum\"\n",
    "# # † = job array based on \"batchnum\" (two rows at a time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A04a_classify_mCT_reads_bismark.pl\n",
    "\n",
    "#!/usr/bin/perl -w\n",
    "use strict;\n",
    "\n",
    "# A04a_classify_mCT_reads_bismark.pl, v0.2 =====================================================\n",
    "# original perl script written by Dr. Chongyuan Luo (@luogenomics)\n",
    "# minor modifications by Choo Liu (@chooliu):\n",
    "# - readability/documentation - minor bug fix (num total mCH over-counted) \n",
    "# inputs: - .sam file from Bismark (compatible with single-end or paired-end alignments)\n",
    "# outputs: - \"_annotations\" .tsv recording each alignment's:\n",
    "#             number of cytosines, mC/C fraction, and call (DNA, RNA, ambiguous)\n",
    "#             this annotation file is subsequently appended to the .bam to keep DNA reads\n",
    "# typical usage: perl A04a_classify_mCT_reads_bismark.pl alignments.sam\n",
    "# =====================================================================================\n",
    "\n",
    "my ($samfile)=($ARGV[0]);\n",
    "my @name; my $name; my @sample; my $um; my $m;\n",
    "my @row; my $xrtag;\n",
    "my $fraction; my $totalch; my @call;\n",
    "\n",
    "# default filtering criteria [*]\n",
    "my $filter_num_CH=3;\n",
    "my $filter_frac_mCHmin=0.5; # DNA (mCH/CH <0.5)\n",
    "my $filter_frac_mCHmax=0.9; # RNA (mCH/CH >0.9)\n",
    "\n",
    "# for each line in .sam file,\n",
    "# count unmethylated and methylated cytosines in CHG and CHH context based on XR-tag \n",
    "\n",
    "# notes:\n",
    "# - skip header header rows (starting with @)\n",
    "# - assumes the XR-tag is in column 15 (bismark output), $row[14]\n",
    "# - counting based on CHG and CHH, indicated with \"h\", \"x\", \"H\", \"X\"\n",
    "#   (does not include \"unknown context\" CN or CHN, marked by \"u\" and \"U\")\n",
    "# - default settings: keep reads with >=3 CHNs and mCHN/CHN fraction <=0.5 [*]\n",
    "#   at present, we only retain call=\"DNA\" and discard all others\n",
    "\n",
    "open sam_in, \"$samfile\" or die $!;\n",
    "open sam_annotations, \">$samfile\\_annotations\" or die $!;\n",
    "\n",
    "while (<sam_in>)\n",
    "{\n",
    "\tchop $_;\n",
    "\tif (substr($_,0,1) eq '@') { print sam_annotations \"\\n\"; }\n",
    "\telse\n",
    "\t{\n",
    "\t\t@row=split(/\\t/,$_);\n",
    "\t\t$xrtag=$row[14];\n",
    "\t\t$um=($xrtag=~tr/hx//); \n",
    "\t\t$m=(($xrtag=~tr/HX//)-1); \n",
    "\t\t$totalch=($um+$m); \n",
    "\t\tif ($totalch==0) { \n",
    "\t\t\t$fraction=999;\n",
    "\t\t\t@call=\"amb\";\n",
    "\t\t} else {\n",
    "\t\t\t$fraction=(($m)/$totalch);\n",
    "\t\t\tif (($totalch>=$filter_num_CH) and ($fraction <= $filter_frac_mCHmin)) { @call=\"DNA\"; }\n",
    "\t\t\t\telsif (($totalch>=$filter_num_CH) and ($fraction>=$filter_frac_mCHmax)) { @call=\"RNA\"; }\n",
    "\t\t\telse { @call=\"amb\"; }\n",
    "\t\t}\n",
    "\t\tprint sam_annotations \"${totalch}\\t${fraction}\\t@call\\n\";\n",
    "\t}\n",
    "}\n",
    "\n",
    "close sam_annotations;\n",
    "close sam_in;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A04a_bismark_map_to_allc.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A04a_bismark.$JOB_ID.$TASK_ID\n",
    "#$ -j y\n",
    "#$ -l h_rt=48:00:00,h_data=16G,highp\n",
    "#$ -pe shared 4\n",
    "#$ -N A04a_bismark\n",
    "#$ -t 1-256\n",
    "#$ -hold_jid_ad A03a_trim\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init ------------------—------------------—-----------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snmCTseq # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "skip_complete=true # <-- for help with incomplete jobs\n",
    "overwrite_partial=true # <-- for help with incomplete jobs\n",
    "\n",
    "# note: estimated time is ~30 min/well so h_rt=48:00:00 may be excessive\n",
    "# alternatives are to use 24hr & resubmit if incomplete or change wells/batch in A01c\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths ------------------—------------------—---------------\n",
    "\n",
    "# helper functions\n",
    "query_metadat () {\n",
    "  awk -F',' -v targetcol=\"$1\" \\\n",
    "      'NR==1 {\n",
    "                for (i=1;i<=NF;i++) {\n",
    "                    if ($i==targetcol) {assayout=i; break} }\n",
    "                print $assayout\n",
    "              }\n",
    "      NR>1 {\n",
    "                print $assayout\n",
    "            }' $metadat_well\n",
    "}\n",
    "\n",
    "# extract target wells, print values for log\n",
    "batchnum=($(query_metadat \"batchnum\"))\n",
    "nwells=${#batchnum[@]}\n",
    "\n",
    "target_well_rows=()\n",
    "for ((row=1; row<=nwells; row++))\n",
    "do\n",
    "    if [[ \"${batchnum[$row]}\" == \"$SGE_TASK_ID\" ]]\n",
    "    then\n",
    "        target_well_rows+=($row)\n",
    "    fi\n",
    "done\n",
    "\n",
    "# filepaths associated with target rows in well-level metadata -----------------\n",
    "# (generally not customizeable because output names set by bismark)\n",
    "\n",
    "wellprefix=($(query_metadat \"wellprefix\"))\n",
    "dir_well=($(query_metadat \"A04a_dir_bismark\"))\n",
    "\n",
    "# .fastqs for input to PE mapping (properly paired read pairs)\n",
    "fastq_r1p=($(query_metadat \"A03a_fqgz_paired_R1\"))\n",
    "fastq_r2p=($(query_metadat \"A03a_fqgz_paired_R2\"))\n",
    "\n",
    "# .fastqs for input to SE mapping, including singletons from trimming & unaligned in PE-mapping\n",
    "fastq_r1singletrim=($(query_metadat \"A03a_fqgz_singletrim_R1\"))\n",
    "fastq_r2singletrim=($(query_metadat \"A03a_fqgz_singletrim_R2\"))\n",
    "fastq_r1unmap=($(query_metadat \"A04a_fqgz_unmap_R1\"))\n",
    "fastq_r2unmap=($(query_metadat \"A04a_fqgz_unmap_R2\"))\n",
    "\n",
    "# (I) bismark mapping .bam output\n",
    "bam_pe=($(query_metadat \"A04a_bam_bismark_PE\"))\n",
    "bam_se1trim=($(query_metadat \"A04a_bam_bismark_SE1trim\"))\n",
    "bam_se2trim=($(query_metadat \"A04a_bam_bismark_SE2trim\"))\n",
    "bam_se1unmap=($(query_metadat \"A04a_bam_bismark_SE1unmap\"))\n",
    "bam_se2unmap=($(query_metadat \"A04a_bam_bismark_SE2unmap\"))\n",
    "\n",
    "\n",
    "\n",
    "# (II) deduplication, (III) read-level filtering, (IV) allc generation ---------\n",
    "# these names customizeable and located at mapping_bismark/$wellprefix[@]/namegivenbelow\n",
    "# (should shift to grabbing from the metadata file for consistency)\n",
    "\n",
    "bam_dedupe_pe=PE_dedupe.bam\n",
    "bam_merge_se=SE_merge.bam\n",
    "bam_mergesort_se=SE_mergesort.bam\n",
    "bam_dedupe_se=SE_dedupe.bam\n",
    "\n",
    "log_picard_pe=picard_PE.log\n",
    "log_picard_se=picard_SE.log\n",
    "\n",
    "sam_dedupeq10_pe=PE_dedupeq10.sam\n",
    "sam_dedupeq10_se=SE_dedupeq10.sam\n",
    "\n",
    "sam_pe_filtered=PE_filt.sam\n",
    "sam_se_filtered=SE_filt.sam\n",
    "\n",
    "bam_pe_filtered=PE_final.bam\n",
    "bam_se_filtered=SE_final.bam\n",
    "\n",
    "allc_PE=PE_allc.tsv.gz\n",
    "allc_SE=SE_allc.tsv.gz\n",
    "\n",
    "allc_final=allc.tsv.gz\n",
    "allctbi_final=allc.tsv.gz.tbi\n",
    "allc_check=allc_check.txt\n",
    "\n",
    "\n",
    "\n",
    "# print target files -----------------------------------------------------------\n",
    "\n",
    "echo \"batch number: $SGE_TASK_ID\"\n",
    "echo \"processing the following rows in well metadata file ($metadat_well):\"\n",
    "for row in ${target_well_rows[@]}\n",
    "    do\n",
    "        echo -e \"$row\\t${wellprefix[$row]}\"\n",
    "    done\n",
    "echo -e \"\\n\\n\"\n",
    "\n",
    "\n",
    "# for each well in batch, apply mC map & quant\n",
    "# (could add check here to skip rows where no trimming output,\n",
    "# but since done by well doesn't cause catastrophic problems)\n",
    "for row in ${target_well_rows[@]} \n",
    "do\n",
    "\n",
    "    # check for existing mapping output\n",
    "    # if final outputs exist, skip; else run mapping .bam\n",
    "    cd $projdir\n",
    "    \n",
    "    if [[ -s mapping_bismark/${wellprefix[$row]}/$allc_final \\\n",
    "        && -s mapping_bismark/${wellprefix[$row]}/$allctbi_final \\\n",
    "        && -s mapping_bismark/${wellprefix[$row]}/$allc_check \\\n",
    "        && \"$skip_complete\"==\"true\" ]]\n",
    "    then\n",
    "        echo -e \"final .allc files for '${wellprefix[$row]}' already exist. skipping this well.'\"\n",
    "    else\n",
    "    \n",
    "        echo -e \"\\n\\napplying bismark to '${wellprefix[$row]}'...\\n\\n\"\n",
    "\n",
    "        # remove old directory if one exists to deal with incomplete files\n",
    "        # albeit the only major issues are .bai and .tbi indices \n",
    "        # (these often are not overwritten by software in the pipeline,\n",
    "        # resulting in \"index is older than file\" errors later on)\n",
    "        if [[ -e mapping_bismark/${wellprefix[$row]} && \"$overwrite_partial\" == \"true\" ]]\n",
    "        then\n",
    "            echo -e \"\\n\\nWARNING: folder for '${wellprefix[$row]}' exists, but not its final allc files.\"\n",
    "            echo \"because overwrite_partial=true, deleting the directory and re-mapping.\"\n",
    "            rm -rf mapping_bismark/${wellprefix[$row]}\n",
    "        fi\n",
    "        \n",
    "        mkdir ${dir_well[$row]}\n",
    "        cd ${dir_well[$row]}\n",
    "        \n",
    "    # (A) run bismark \"two-stage\" mapping -------------------------------------\n",
    "    # in: .fastqs from trimming: four .fastqs,\n",
    "    #     properly paired ($fastq_r2p, $fastq_r1p) and trimming singletons\n",
    "    #    ($fastq_r1singletrim, $fastq_r2singletrim)\n",
    "    # out: - paired-end alignments out ($bam_pe, $bam_pe_unmap1, $bam_pe_unmap2)\n",
    "    #      - single-end .bam alignments out ($bam_single1, $bam_single2)\n",
    "    #      - key log files (e.g., mapping rate) \n",
    "    # -------------------------------------------------------------------------\n",
    "\n",
    "       # (Ai.) PE [2 to 10 minutes]\n",
    "       # assumptions: pairs that map ambiguously in paired-end mode should be discarded\n",
    "       \n",
    "       bismark $ref_dir --bowtie2 --pbat \\\n",
    "                -1 $projdir/${fastq_r1p[$row]} -2 $projdir/${fastq_r2p[$row]} \\\n",
    "                --maxins 2000 --multicore 3 --un --ambiguous\n",
    "\n",
    "        # (Aii.) single-end, R1 [1 to 3 minutes]\n",
    "        bismark $ref_dir --bowtie2 --pbat -se \\\n",
    "                $projdir/${fastq_r1unmap[$row]}:$projdir/${fastq_r1singletrim[$row]} --multicore 3\n",
    "\n",
    "        # (Aiii.) single-end, R2 [1 to 3 minutes]\n",
    "        bismark $ref_dir --bowtie2 -se \\\n",
    "                $projdir/${fastq_r2unmap[$row]}:$projdir/${fastq_r2singletrim[$row]} --multicore 3\n",
    "\n",
    "        # cleanup\n",
    "        rm *ambiguous_reads*\n",
    "        \n",
    "\n",
    "    # (B) de-deuplication -----------------------------------------------------\n",
    "    # in: paired-end .bam (name sorted), four single-end .bam (unsorted, need to sort)\n",
    "    # out: de-duplicated PE .bam, de-duplicated combined SE .bam\n",
    "    # -------------------------------------------------------------------------\n",
    "        \n",
    "        # (Bi.) PE [<1 min]\n",
    "        picard MarkDuplicates -I $(basename ${bam_pe[$row]}) \\\n",
    "            --ASSUME_SORT_ORDER \"queryname\" --OPTICAL_DUPLICATE_PIXEL_DISTANCE 2500 \\\n",
    "            --ADD_PG_TAG_TO_READS false --REMOVE_DUPLICATES \\\n",
    "            -O $bam_dedupe_pe -M $log_picard_pe\n",
    "\n",
    "        # (Bii.) SE [1-2 min] \n",
    "        samtools merge $bam_merge_se \\\n",
    "            $(basename ${bam_se1trim[$row]}) $(basename ${bam_se2trim[$row]}) \\\n",
    "            $(basename ${bam_se1unmap[$row]}) $(basename ${bam_se2unmap[$row]})\n",
    "            \n",
    "        samtools sort $bam_merge_se > $bam_mergesort_se\n",
    "        \n",
    "        picard MarkDuplicates -I $bam_mergesort_se \\\n",
    "            --OPTICAL_DUPLICATE_PIXEL_DISTANCE 2500 \\\n",
    "            --ADD_PG_TAG_TO_READS false --REMOVE_DUPLICATES \\\n",
    "            -O $bam_dedupe_se -M $log_picard_se\n",
    "\n",
    "        # (Biii.) clean-up\n",
    "        rm $bam_merge_se $bam_mergesort_se\n",
    "\n",
    "\n",
    "    # (C) read-level filtering ------------------------------------------------\n",
    "    # via MAPQ, mCH levels to remove ambig/RNA reads\n",
    "    # MAPQ >= 10 & .sam conversion --> then perl script to output annotations\n",
    "    # -------------------------------------------------------------------------\n",
    "\n",
    "        # (Ci.) PE [1-3 min]\n",
    "        # this assumes *any* read meeting the filtering criteria is kept;\n",
    "        # if both R1 & R2 of pair need to pass filtering criteria (AND instead of OR),\n",
    "        #     needs an added annotations --> wide step like the below\n",
    "        # sed '$!N;s/\\n/ /' ${sam_dedupe_pe_q10}_annotations \\\n",
    "        #     | awk '{print ( ($3 == \"DNA\") && ($6 == \"DNA\") )}' \\\n",
    "        #     | awk '{print $0}1' > ${sam_dedupe_pe_q10}_annotations_bothpairs\n",
    "        # then awk filter off of this 'bothpairs' file\n",
    "\n",
    "        samtools view -hq 10 $bam_dedupe_pe > $sam_dedupeq10_pe\n",
    "        perl $projdir/Scripts/A04a_classify_mCT_reads_bismark.pl $sam_dedupeq10_pe\n",
    "        awk 'NR == FNR { if ($0==\"\" || $3==\"DNA\") line[NR]; next } (FNR in line)' \\\n",
    "            ${sam_dedupeq10_pe}_annotations ${sam_dedupeq10_pe} |\n",
    "            samtools view -b - | samtools sort - > $bam_pe_filtered\n",
    "        samtools index $bam_pe_filtered\n",
    "        \n",
    "        # (Cii.) SE [1-3 min]\n",
    "        samtools view -hq 10 $bam_dedupe_se > $sam_dedupeq10_se\n",
    "        perl $projdir/Scripts/A04a_classify_mCT_reads_bismark.pl $sam_dedupeq10_se\n",
    "        awk 'NR == FNR { if ($0==\"\" || $3==\"DNA\") line[NR]; next } (FNR in line)' \\\n",
    "            ${sam_dedupeq10_se}_annotations ${sam_dedupeq10_se} |\n",
    "            samtools view -b - | samtools sort - > $bam_se_filtered\n",
    "        samtools index $bam_se_filtered\n",
    "\n",
    "        # (Ciii.) optional clean-up (comment out as desired)\n",
    "        rm $bam_dedupe_pe $sam_dedupeq10_pe\n",
    "        rm $bam_dedupe_se $sam_dedupeq10_se\n",
    "        # rm ${sam_dedupeq10_pe}_annotations ${sam_dedupeq10_se}_annotations\n",
    "        \n",
    "\n",
    "    # (D) generate .allc tsv --------------------------------------------------\n",
    "    # in: SE and PE .bam files\n",
    "    # out: tab-separated file with chr, pos, seq context, methyl reads, total cov\n",
    "    #      generated separatedly for PE and SE, then joined\n",
    "    #      (for historical paired-end dev, but might be faster to join .bam first)\n",
    "    # -------------------------------------------------------------------------\n",
    "\n",
    "        # (Di.) make separate PE, SE allc files [~3-10 minutes] \n",
    "        allcools bam-to-allc -bam $bam_pe_filtered --cpu 1 \\\n",
    "            --reference_fasta $ref_fasta --output_path $allc_PE --convert_bam_strandness\n",
    "        allcools bam-to-allc -bam $bam_se_filtered --cpu 1 \\\n",
    "            --reference_fasta $ref_fasta --output_path $allc_SE --convert_bam_strandness\n",
    "\n",
    "        # (Dii.) merge the .allc files into one [~5 min]\n",
    "        # before .allc generation --> .allc\n",
    "        allcools merge-allc --allc_paths $allc_PE $allc_SE --cpu 1 \\\n",
    "                --output_path $allc_final \\\n",
    "                --chrom_size_path $ref_chromsizes\n",
    "                \n",
    "        # (Diii.) rough check if allc finished, can check for warnings via grep \n",
    "        # main pitfall: bam_to_allc & merge-allc can silently fail from mem/threading issues,\n",
    "        # resulting in invalid last line or not all chromosomes being displayed \n",
    "        \n",
    "        du -h *allc* > $allc_check # PE + SE should approx add up to final\n",
    "        gunzip -c $allc_final | awk -F '\\t' '{print $1}' | uniq -c >> $allc_check\n",
    "        \n",
    "        # usual file size >20Mb\n",
    "        if [[ $(( $(stat -c%s $bam_pe_filtered) + $(stat -c%s $bam_se_filtered) < 3000000 )) == 1 ]]\n",
    "        then\n",
    "            echo -e \"\\n\\nWARNING: .bam file sizes unusually small (<3Mb).\" \n",
    "        fi >> $allc_check\n",
    "        \n",
    "        # comment out or change to last chr depending on ref chromosome used\n",
    "        if [[ $(tail $allc_check | grep 'KI' | wc -l) == 0 ]]\n",
    "        then\n",
    "            echo \"WARNING: KI scaffolds missing; merge .all.gz terminated early? (or low coverage)\"\n",
    "        fi >> $allc_check\n",
    "        \n",
    "        # with mem issue the .tsv may fail mid-writing, so have incorrect # cols\n",
    "        if [[ $( gunzip -c $allc_final | awk '{print NF}' | uniq | wc -l) != 1 ]]\n",
    "            then\n",
    "            echo \"WARNING: different numbers of columns per row detected?\"\n",
    "            fi >> $allc_check\n",
    "            \n",
    "        # (Vc.) optional cleanup\n",
    "        # could also delete pre-filtering alignment files if not using\n",
    "        rm $allc_PE $allc_PE.tbi $allc_SE $allc_SE.tbi\n",
    "    \n",
    "\n",
    "fi\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A04a_bismark' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `date `\n",
    "echo \" \"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A04b_check_bismark.sub\n",
    "\n",
    "# #!/bin/bash\n",
    "# #$ -cwd\n",
    "# #$ -o sublogs/A04b_check_bismark.$JOB_ID\n",
    "# #$ -j y\n",
    "# #$ -l h_rt=2:00:00,h_data=4G\n",
    "# #$ -N A04b_bischeck\n",
    "# #$ -hold_jid A04a_bismark\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init ------------------—------------------—-----------------------\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths ------------------—------------------—---------------\n",
    "\n",
    "query_metadat () {\n",
    "  awk -F',' -v targetcol=\"$1\" \\\n",
    "      'NR==1 {\n",
    "                for (i=1;i<=NF;i++) {\n",
    "                    if ($i==targetcol) {assayout=i; break} }\n",
    "                print $assayout\n",
    "              } \n",
    "      NR>1 {\n",
    "                print $assayout\n",
    "            }' $metadat_well\n",
    "}\n",
    "\n",
    "check_filepaths_in_assay() {\n",
    "    for file in $@\n",
    "        do \n",
    "        if [[ ! -s $file ]]\n",
    "            then\n",
    "                echo \"missing '$file'\"\n",
    "            fi\n",
    "        done\n",
    "}\n",
    "\n",
    "check_filepath_in_batch() {\n",
    "target_array=$@\n",
    "for ((target_batch=1; target_batch<=$nbatches; target_batch++))\n",
    "    do\n",
    "        target_well_rows=()\n",
    "        for ((row=1; row<=nwells; row++))\n",
    "        do\n",
    "            if [[ \"${batchnum[$row]}\" == \"$target_batch\" ]]\n",
    "            then\n",
    "                target_well_rows+=($row)\n",
    "            fi\n",
    "        done\n",
    "        \n",
    "        batch_allc_list=()\n",
    "        for row in ${target_well_rows[@]}\n",
    "        do\n",
    "            batch_allc_list+=(${allctbi[$row]})\n",
    "        done\n",
    "        \n",
    "        num_allc_missing=$(check_filepaths_in_assay ${batch_allc_list[@]} | wc -l)\n",
    "        echo -e \"${target_batch} \\t $num_allc_missing\"\n",
    "\n",
    "    done \\\n",
    "    | awk -F \"\\t\" '($2 > 0) { print $0 }'\n",
    "}\n",
    "\n",
    "batchnum=($(query_metadat \"batchnum\"))\n",
    "\n",
    "nwells=${#batchnum[@]}\n",
    "nbatches=${batchnum[-1]}\n",
    "\n",
    "\n",
    "\n",
    "# apply checks for A04a output ------------------—------------------—-----------\n",
    "\n",
    "echo \"-----------------------------------------------------------------\"\n",
    "echo \"A. printing number of .allc files missing (by batch)... \"\n",
    "echo \"-----------------------------------------------------------------\"\n",
    "\n",
    "allcgz=($(query_metadat \"A04a_allc_final\"))\n",
    "allctbi=($(query_metadat \"A04a_allctbi_final\"))\n",
    "allctxt=($(query_metadat \"A04a_txt_allccheck\"))\n",
    "\n",
    "echo \"checking allc.tsv.gz\"\n",
    "echo -e \"batchnum\\tnum_missing\"\n",
    "check_filepath_in_batch ${allcgz[@]}\n",
    "\n",
    "echo \"checking allc.tsv.gz.tbi\"\n",
    "echo -e \"batchnum\\tnum_missing\"\n",
    "check_filepath_in_batch ${allctbi[@]}\n",
    "\n",
    "echo \"checking allc_check.txt\"\n",
    "echo -e \"batchnum\\tnum_missing\"\n",
    "check_filepath_in_batch ${allctxt[@]}\n",
    "\n",
    "echo -e \"\\n\\nsuggest re-running and checking sublog output of above batches.\"\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n-----------------------------------------------------------------\"\n",
    "echo \"B. checking each expected file (from $metadat_well)\"\n",
    "echo -e \"-----------------------------------------------------------------\\n\"\n",
    "\n",
    "echo -e \"\\nchecking .allc.gz files:\\n\"\n",
    "check_filepaths_in_assay ${allcgz[@]}\n",
    "\n",
    "echo -e \"\\nchecking .allc.gz.tbi files:\\n\"\n",
    "check_filepaths_in_assay ${allctbi[@]}\n",
    "\n",
    "echo -e \"\\nchecking allc_check.txt files:\\n\"\n",
    "check_filepaths_in_assay ${allctxt[@]}\n",
    "\n",
    "echo -e \"\\n* checks the A04a output columns of 'metadat_well' if the file exists and is non-empty.\"\n",
    "echo \"* if none missing, will only output target column names above.\"\n",
    "echo \"* if some declared 'missing' but all other checks OK, may just be no/few reads surviving trimming.\"\n",
    "echo \"  (check 'fastq_demultip/' and associated fastp logs e.g., fastq_trimmed/wellprefix.html report)\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo \"-----------------------------------------------------------------\"\n",
    "echo \"C. checking list of .allc.gz.tbi files (by plate)...\"\n",
    "echo \"-----------------------------------------------------------------\"\n",
    "\n",
    "\n",
    "echo -e \"\\nnumber of .allc.gz files found:\\n\"\n",
    "prefix_num=0\n",
    "for prefix in $(cut -f 2 -d ',' $metadat_plate | tail -n +2);\n",
    "    do\n",
    "        let \"prefix_num+=1\"\n",
    "        find mapping_bismark/ -maxdepth 2 -mindepth 2  -wholename \"*${prefix}*/allc.tsv.gz\" \\\n",
    "            | awk -F '\\\\/' '{print $2 \"\\t\" $0}' > Scripts/A04b_allclists_${prefix_num}.tsv\n",
    "        nallc_plate=$(wc -l Scripts/A04b_allclists_${prefix_num}.tsv | awk -F \" \" '{print $1}')\n",
    "        echo \"$prefix : $nallc_plate\"\n",
    "        let \"ntot_allc+=$nallc_plate\"\n",
    "done\n",
    "\n",
    "echo -e \"\\nnumber of .allc.gz.tbi files found:\\n\"\n",
    "for prefix in $(cut -f 2 -d ',' $metadat_plate | tail -n +2);\n",
    "    do\n",
    "        nallctbi_plate=$(\n",
    "            find mapping_bismark/ -maxdepth 2 -mindepth 2  -wholename \"*${prefix}*/allc_check.txt\" | wc -l)\n",
    "        echo \"$prefix : $nallctbi_plate\"\n",
    "        let \"ntot_allctbi+=$nallctbi_plate\"\n",
    "done\n",
    "\n",
    "echo -e \"\\nnumber of allc_check.txt files found:\\n\"\n",
    "for prefix in $(cut -f 2 -d ',' $metadat_plate | tail -n +2);\n",
    "    do\n",
    "        nallcchecktxt_plate=$(\n",
    "            find mapping_bismark/ -maxdepth 2 -mindepth 2  -wholename \"*${prefix}*/allc.tsv.gz.tbi\" | wc -l)\n",
    "        echo \"$prefix : $nallcchecktxt_plate\"\n",
    "        let \"ntot_allctxt+=$nallcchecktxt_plate\"\n",
    "done\n",
    "\n",
    "echo -e \"\\ncompare to the number of trimmed R1 .fastq.gz files found...:\\n\"\n",
    "for prefix in $(cut -f 2 -d ',' $metadat_plate | tail -n +2);\n",
    "    do\n",
    "        ntrimmed_plate=$(\n",
    "            find fastq_trimmed/ -maxdepth 1 -mindepth 1  -wholename \"*${prefix}*paired_R1.fastq.gz\" | wc -l)\n",
    "        echo \"$prefix : $ntrimmed_plate\"\n",
    "        let \"ntot_allctxt+=$ntrimmed_plate\"\n",
    "done\n",
    "\n",
    "\n",
    "echo \"* expected same # .allc as of # trimmed .fastq.gz files input into bismark mapping.\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n-----------------------------------------------------------------\"\n",
    "echo \"C. checking log files for issues.\"\n",
    "echo -e \"-----------------------------------------------------------------\\n\"\n",
    "\n",
    "for txtfile in ${allctxt[@]}\n",
    "    do\n",
    "        if [[ -s $txtfile ]]\n",
    "        then\n",
    "            if $(grep -q 'WARNING' $txtfile && grep -vq 'file sizes' $txtfile)\n",
    "            then \n",
    "                echo \"warning in '$txtfile' ?\"\n",
    "            fi\n",
    "        fi\n",
    "    done\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n-----------------------------------------------------------------\"\n",
    "echo \"C. checking log files for issues.\"\n",
    "echo -e \"-----------------------------------------------------------------\\n\"\n",
    "\n",
    "\n",
    "\n",
    "echo \"checking if 'completed' in sublogs/A04a_bismark* output.\"\n",
    "echo \"if any filename is printed, the associated batch may have not completed mapping.\"\n",
    "\n",
    "grep -c 'A04a_bismark completed' sublogs/A04b_bismark* | awk -F \":\" '$2==0 {print $1}'\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A04b_bischeck' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID ended on:   \" `date `\n",
    "echo \" \"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A04c_allc_to_mcds.sub\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A04c_makemcds.$JOB_ID.$TASK_ID\n",
    "#$ -j y\n",
    "#$ -l h_rt=24:00:00,h_data=8G,highp\n",
    "#$ -N A04c_mcdsgen\n",
    "#$ -t 1-16\n",
    "#$ -pe shared 8\n",
    "#$ -hold_jid A04b_bischeck\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init ------------------—------------------—-----------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snmCTseq # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "\n",
    "\n",
    "# run mcds generation ------------------—------------------—--------------------\n",
    "\n",
    "# note: the .allc files are aggregated into the mcds\n",
    "# the mcds is not a single file, but more like a directory with binary compression\n",
    "\n",
    "mkdir mcds\n",
    "\n",
    "allcools generate-dataset  \\\n",
    "    --allc_table Scripts/A04b_allclists_${SGE_TASK_ID}.tsv \\\n",
    "    --output_path mcds/$SGE_TASK_ID.mcds \\\n",
    "    --chrom_size_path $ref_chromsizes \\\n",
    "    --obs_dim cell \\\n",
    "    --cpu 8 \\\n",
    "    --chunk_size 400 \\\n",
    "    --regions chrom100k 100000 \\\n",
    "    --regions genebody $ref_genebody \\\n",
    "    --quantifiers chrom100k count CGN,CHN \\\n",
    "    --quantifiers genebody count CGN,CHN\n",
    "\n",
    "# not standard, but but consider adding:\n",
    "#     --regions chrom5k 5000 \\\n",
    "#     --regions geneslop2k $ref_geneslop2k \\\n",
    "# (genebody +/- 2kb to include promoter region + increase cov)\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A04c_allc_to_mcds' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `date `\n",
    "echo \" \"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A04d_calc_mC_fracs.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A04d_mCfracs.$JOB_ID.$TASK_ID\n",
    "#$ -j y\n",
    "#$ -l h_rt=8:00:00,h_data=4G\n",
    "#$ -N A04d_mCfracs\n",
    "#$ -t 1-256\n",
    "#$ -hold_jid A04a_bismark\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init ------------------—------------------—-----------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "action_metadata_exists=\"overwrite\" # <-- for help with incomplete jobs; overwrite, append, rename\n",
    "check_lambda=false # <-- lambda phage spike-in (currently we use only in snmC-seq2/3)\n",
    "check_autosomal_only=false # <-- exclude mitochondrial, scaffolds, lambda\n",
    "\n",
    "# output file\n",
    "# (should probably make these explicitly plate-named vs numbered in future)\n",
    "metadat_out=Metadata/A04d_mCfrac_${SGE_TASK_ID}.tsv\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths ------------------—------------------—---------------\n",
    "\n",
    "# helper functions\n",
    "query_metadat () {\n",
    "  awk -F',' -v targetcol=\"$1\" \\\n",
    "      'NR==1 {\n",
    "                for (i=1;i<=NF;i++) {\n",
    "                    if ($i==targetcol) {assayout=i; break} }\n",
    "                print $assayout\n",
    "              }\n",
    "      NR>1 {\n",
    "                print $assayout\n",
    "            }' $metadat_well\n",
    "}\n",
    "\n",
    "# extract target wells, print values for log\n",
    "batchnum=($(query_metadat \"batchnum\"))\n",
    "nwells=${#batchnum[@]}\n",
    "\n",
    "target_well_rows=()\n",
    "for ((row=1; row<=nwells; row++))\n",
    "do\n",
    "    if [[ \"${batchnum[$row]}\" == \"$SGE_TASK_ID\" ]]\n",
    "    then\n",
    "        target_well_rows+=($row)\n",
    "    fi\n",
    "done\n",
    "\n",
    "wellprefix=($(query_metadat \"wellprefix\"))\n",
    "allcfile=($(query_metadat \"A04a_allc_final\"))\n",
    "\n",
    "# for methylation fraction\n",
    "calc_mC_frac () {\n",
    "    awk -v context=$1 '$4 ~ context' ${2:-$tmpfile} |\n",
    "        awk '{mC+=$5; tot+=$6} END {\n",
    "            if (tot == 0) { print \"0\\t0\\tNA\" } else { print mC \"\\t\" tot \"\\t\" mC/tot }}'\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "# print target files ------------------—------------------—---------------------\n",
    "\n",
    "echo \"batch number: $SGE_TASK_ID\"\n",
    "echo \"processing the following rows in well metadata file ($metadat_well):\"\n",
    "for row in ${target_well_rows[@]}\n",
    "    do\n",
    "        echo -e \"$row\\t${wellprefix[$row]}\"\n",
    "    done\n",
    "echo -e \"\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "# checking for existing metadata file ------------------—------------------—----\n",
    "\n",
    "# columns are wellprefix, then for a given sequence context\n",
    "# num reads supporting methylation, total coverage, then fraction methylated\n",
    "\n",
    "header=\"wellprefix\\tmLam\\tLam\\tmLamfrac\\tmCCC\\tCCC\\tmCCCfrac\\tmCG\\tCG\\tmCGfrac\\tmCH\\tCH\\tmCHfrac\\n\"\n",
    "\n",
    "if [[ ! -e $metadat_out ]]\n",
    "then\n",
    "    printf $header > $metadat_out\n",
    "else\n",
    "    echo \"WARNING: $metadat_out seems to already exist.\"\n",
    "    if [[ $action_metadata_exists == \"overwrite\" ]]\n",
    "        then\n",
    "        echo \"overwriting the existing file. (action_metadata_exists=='overwrite').\"\n",
    "        cat \"$header\" > $metadat_out\n",
    "    elif [[ $action_metadata_exists == \"append\" ]]\n",
    "        then\n",
    "            echo \"appending to the existing file. (action_metadata_exists=='append')\"\n",
    "    elif [[ $action_metadata_exists == \"rename\" ]]\n",
    "        then\n",
    "            metadat_out=Metadata/A04d_mCfrac_${SGE_TASK_ID}_B.tsv\n",
    "            echo \"renaming output file to $metadat_out to avoid overwriting (action_metadata_exists=='rename').\"\n",
    "            cat \"$header\" > $metadat_out\n",
    "    else\n",
    "        echo \"exiting. (check 'action_metadata_exists' variable if to change action.)\"\n",
    "        exit 1\n",
    "    fi\n",
    "fi\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# loop through allc, calculate methylation fracs ------------------—------------\n",
    "# assuming methylation given for CHN or CGN contexts\n",
    "# (usually <1 minute per file)\n",
    "\n",
    "for row in ${target_well_rows[@]} \n",
    "do\n",
    "\n",
    "    # check for existing mapping output\n",
    "    # if final outputs exist, skip; else run awk tabulization\n",
    "    cd $projdir\n",
    "    \n",
    "    if [[ ! -s mapping_bismark/${wellprefix[$row]}/$allc_final \\\n",
    "        && ! -s mapping_bismark/${wellprefix[$row]}/$allctbi_final ]]\n",
    "    then\n",
    "        echo -e \"error: final .allc files for '${wellprefix[$row]}' missing. skipping this well.'\"\n",
    "    else\n",
    "    \n",
    "    echo \"processing '${wellprefix[$row]}'...\"\n",
    "        \n",
    "    # create temporary unzipped .allc file to awk through\n",
    "    # (if there are \"tmp_*\" files in '$projdir/Metadata', this script may have failed/timeout)\n",
    "    tmpfile=\"Metadata/tmp_${wellprefix[$row]}\"\n",
    "    if [[ \"$check_lambda\" == \"true\" ]]\n",
    "    then\n",
    "        # use a partially methylated lambda, methylated at all but CAG, CTG\n",
    "        gunzip -c ${allcfile[$row]} | awk '$1 == \"chrL\" && $4 != \"CAG\" && $4 != \"CTG\"' \\\n",
    "                > ${tmpfile}_lambda\n",
    "        mlambda=$(calc_mC_frac \"^[ACTG][ACTG][ACTG]\" ${tmpfile}_lambda )\n",
    "        rm ${tmpfile}_lambda\n",
    "    else\n",
    "        mlambda=\"0\\t0\\tNA\"\n",
    "    fi\n",
    "    \n",
    "    if [[ \"$check_autosomal_only\" == \"true\" ]]\n",
    "        then\n",
    "            gunzip -c ${allcfile[$row]} | awk '$1 ~ \"^chr[0-9]\"' > $tmpfile\n",
    "        else\n",
    "            gunzip -c ${allcfile[$row]} | awk '$1 != \"chrL\"' > $tmpfile\n",
    "    fi\n",
    "\n",
    "    # calculate CCC, CGN, CHN\n",
    "    mccc=$( calc_mC_frac \"^CCC\" )\n",
    "    mcg=$( calc_mC_frac \"^CG[ACTG]\" )\n",
    "    mch=$( calc_mC_frac \"^C[ACT][ACTG]\" )\n",
    "\n",
    "    # record metrics, remove tmp file\n",
    "    echo -e \"${wellprefix[$row]}\\t${mlambda}\\t${mccc}\\t${mcg}\\t${mch}\\n\" >> $metadat_out\n",
    "    rm $tmpfile\n",
    "\n",
    "fi\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A04d_calc_mC_fracs' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `date `\n",
    "echo \" \"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A04e_samstats_DNA.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A04e_samstat_DNA.$JOB_ID.$TASK_ID\n",
    "#$ -j y\n",
    "#$ -l h_rt=6:00:00,h_data=24G\n",
    "#$ -N A04e_samstat_DNA\n",
    "#$ -t 1-256\n",
    "#$ -hold_jid_ad A04a_bismark\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init ------------------—------------------—-----------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snmCTseq # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "skip_complete=true # <-- for help with incomplete jobs\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths ------------------—------------------—---------------\n",
    "\n",
    "# helper functions\n",
    "query_metadat () {\n",
    "  awk -F',' -v targetcol=\"$1\" \\\n",
    "      'NR==1 {\n",
    "                for (i=1;i<=NF;i++) {\n",
    "                    if ($i==targetcol) {assayout=i; break} }\n",
    "                print $assayout\n",
    "              }\n",
    "      NR>1 {\n",
    "                print $assayout\n",
    "            }' $metadat_well\n",
    "}\n",
    "\n",
    "# extract target wells, print values for log\n",
    "\n",
    "batchnum=($(query_metadat \"batchnum\"))\n",
    "nwells=${#batchnum[@]}\n",
    "\n",
    "target_well_rows=()\n",
    "for ((row=1; row<=nwells; row++))\n",
    "do\n",
    "    if [[ \"${batchnum[$row]}\" == \"$SGE_TASK_ID\" ]]\n",
    "    then\n",
    "        target_well_rows+=($row)\n",
    "    fi\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "# filepaths associated with target rows in well-level metadata -----------------\n",
    "\n",
    "wellprefix=($(query_metadat \"wellprefix\"))\n",
    "dir_well=($(query_metadat \"A04a_dir_bismark\"))\n",
    "\n",
    "\n",
    "# samtools stats on each well in the batch ------------------—------------------\n",
    "\n",
    "for row in ${target_well_rows[@]} \n",
    "do\n",
    "\n",
    "    cd $projdir\n",
    "    \n",
    "    if [[ -s ${dir_well[$row]}/samstats_PE \\\n",
    "        && -s ${dir_well[$row]}/samstats_SE \\\n",
    "        && \"$skip_complete\"==\"true\" ]]\n",
    "    then\n",
    "        echo -e \"samstats output for '${wellprefix[$row]}' already exists. skipping this well.'\"\n",
    "    else\n",
    "    \n",
    "        echo -e \"\\n\\nprofiling .bams from '${wellprefix[$row]}'...\\n\\n\"\n",
    "\n",
    "        cd ${dir_well[$row]}\n",
    "        \n",
    "        # run samtools stats\n",
    "        samtools stats PE_final.bam | grep '^SN' | cut -f 2,3 > samstats_PE\n",
    "        samtools stats SE_final.bam | grep '^SN' | cut -f 2,3 > samstats_SE\n",
    "            \n",
    "    fi\n",
    "done\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A04e_samstats_DNA' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A04f_coverage_DNA.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A04f_coverage_DNA.$JOB_ID.$TASK_ID\n",
    "#$ -j y\n",
    "#$ -l h_rt=6:00:00,h_data=24G\n",
    "#$ -N A04f_coverage_DNA\n",
    "#$ -t 1-256\n",
    "#$ -hold_jid_ad A04a_bismark\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "# environment init ------------------—------------------—-----------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snmCTseq # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "\n",
    "skip_complete=true # <-- for help with incomplete jobs\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths ------------------—------------------—---------------\n",
    "\n",
    "# helper functions\n",
    "query_metadat () {\n",
    "  awk -F',' -v targetcol=\"$1\" \\\n",
    "      'NR==1 {\n",
    "                for (i=1;i<=NF;i++) {\n",
    "                    if ($i==targetcol) {assayout=i; break} }\n",
    "                print $assayout\n",
    "              }\n",
    "      NR>1 {\n",
    "                print $assayout\n",
    "            }' $metadat_well\n",
    "}\n",
    "\n",
    "# extract target wells, print values for log\n",
    "\n",
    "batchnum=($(query_metadat \"batchnum\"))\n",
    "nwells=${#batchnum[@]}\n",
    "\n",
    "target_well_rows=()\n",
    "for ((row=1; row<=nwells; row++))\n",
    "do\n",
    "    if [[ \"${batchnum[$row]}\" == \"$SGE_TASK_ID\" ]]\n",
    "    then\n",
    "        target_well_rows+=($row)\n",
    "    fi\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "# filepaths associated with target rows in well-level metadata ---------------\n",
    "\n",
    "wellprefix=($(query_metadat \"wellprefix\"))\n",
    "dir_well=($(query_metadat \"A04a_dir_bismark\"))\n",
    "\n",
    "\n",
    "\n",
    "# run coverage check on each well in the batch ------------------—--------------\n",
    "\n",
    "for row in ${target_well_rows[@]} \n",
    "do\n",
    "\n",
    "    cd $projdir\n",
    "    \n",
    "    if [[ -s ${dir_well[$row]}/total_cov_by_chr \\\n",
    "        && -s ${dir_well[$row]}/nbases_cov_by_chr \\\n",
    "        && \"$skip_complete\"==\"true\" ]]\n",
    "    then\n",
    "        echo -e \"final coverage files for '${wellprefix[$row]}' already exist. skipping this well.'\"\n",
    "    else\n",
    "    \n",
    "        echo -e \"\\n\\nprofiling .bam coverage from '${wellprefix[$row]}'...\\n\\n\"\n",
    "\n",
    "        cd ${dir_well[$row]}\n",
    "        \n",
    "        # use samtools mpileup for total coverage\n",
    "        # across combined paired- and single-end files (easiest to join beforehand)\n",
    "        samtools merge tmp_PEplusSE.bam PE_final.bam SE_final.bam\n",
    "        samtools mpileup tmp_PEplusSE.bam | cut -f 1,4 > tmp_coverage_mpileup\n",
    "\n",
    "        # aggregate by chromosome\n",
    "        # (useful for sex-checks)\n",
    "        cut -f 1 tmp_coverage_mpileup | uniq -c > nbases_cov_by_chr\n",
    "        awk '{covsums[$1]+=$2} END {for (key in covsums) printf(\"%s\\t%s\\n\", key, covsums[key])}' \\\n",
    "            tmp_coverage_mpileup > total_cov_by_chr\n",
    "        rm tmp_PEplusSE.bam tmp_coverage_mpileup\n",
    "            \n",
    "    fi\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A04f_coverage_DNA' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
