{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## overall commands\n",
    "# ## might be cleaner to wait for STAR mapping to finish 100% (run A04a only --> check output)\n",
    "# ## in lieu of submitting all of the below, but technically could run all cmds at once\n",
    "    \n",
    "# # * = job array based on \"platenum\"\n",
    "# # † = job array based on \"batchnum\" (two rows at a time)\n",
    "\n",
    "# qsub Scripts/A05a_star_mapping.sub # † \n",
    "# qsub Scripts/A05b_check_star.sub \n",
    "# qsub Scripts/A05c_star_filtering.sub # † \n",
    "# qsub Scripts/A05d_featurecounts.sub   # *\n",
    "# qsub Scripts/A05e_star_bam_stats.sub # † \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A05a_star_mapping.sub\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A05a_star.$JOB_ID.$TASK_ID\n",
    "#$ -j y\n",
    "#$ -l h_rt=8:00:00,h_data=16G,exclusive\n",
    "#$ -pe shared 4\n",
    "#$ -N A05a_star\n",
    "#$ -t 1-256\n",
    "#$ -hold_jid_ad A03a_trim\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "# environment init ------------------—------------------—-----------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snmCTseq # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "skip_complete=true # <-- for help with incomplete jobs\n",
    "overwrite_partial=true # <-- for help with incomplete jobs\n",
    "\n",
    "# (most of the) STAR settings\n",
    "# originally adapted from ENCODE guidelines\n",
    "star_params=\"--runThreadN 4 --genomeDir ${ref_starfolder} --genomeLoad LoadAndKeep \\\n",
    "–alignEndsType EndToEnd --outSAMtype BAM Unsorted \\\n",
    "--outSAMattributes NH HI AS NM MD --outSAMstrandField intronMotif \\\n",
    "--sjdbOverhang 149 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 \\\n",
    "--alignSJDBoverhangMin 1 –outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 \\\n",
    "--alignIntronMax 1000000 --alignMatesGapMax 1000000 --readFilesCommand zcat \"\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths ------------------—------------------—---------------\n",
    "\n",
    "# helper functions\n",
    "query_metadat () {\n",
    "  awk -F',' -v targetcol=\"$1\" \\\n",
    "      'NR==1 {\n",
    "                for (i=1;i<=NF;i++) {\n",
    "                    if ($i==targetcol) {assayout=i; break} }\n",
    "                print $assayout\n",
    "              }\n",
    "      NR>1 {\n",
    "                print $assayout\n",
    "            }' $metadat_well\n",
    "}\n",
    "\n",
    "# extract target wells, print values for log\n",
    "batchnum=($(query_metadat \"batchnum\"))\n",
    "nwells=${#batchnum[@]}\n",
    "\n",
    "target_well_rows=()\n",
    "for ((row=1; row<=nwells; row++))\n",
    "do\n",
    "    if [[ \"${batchnum[$row]}\" == \"$SGE_TASK_ID\" ]]\n",
    "    then\n",
    "        target_well_rows+=($row)\n",
    "    fi\n",
    "done\n",
    "\n",
    "# filepaths associated with target rows in well-level metadata -----------------\n",
    "\n",
    "wellprefix=($(query_metadat \"wellprefix\"))\n",
    "dir_well=($(query_metadat \"A05a_dir_star\"))\n",
    "\n",
    "# .fastqs for input to PE mapping (properly paired read pairs)\n",
    "fastq_r1p=($(query_metadat \"A03a_fqgz_paired_R1\"))\n",
    "fastq_r2p=($(query_metadat \"A03a_fqgz_paired_R2\"))\n",
    "\n",
    "# .fastqs for input to SE mapping, including singletons from trimming & unaligned in PE-mapping\n",
    "fastq_r1singletrim=($(query_metadat \"A03a_fqgz_singletrim_R1\"))\n",
    "fastq_r2singletrim=($(query_metadat \"A03a_fqgz_singletrim_R2\"))\n",
    "\n",
    "# temporary/intermediate mapping files -----------------------------------------------\n",
    "# (expected output generated by STAR, given outname prefixes PE., SE1., SE2.)\n",
    "\n",
    "fastq_pe_unmap1=PE.Unmapped.out.mate1\n",
    "fastq_pe_unmap2=PE.Unmapped.out.mate2\n",
    "\n",
    "bam_pe=PE.Aligned.out.bam\n",
    "bam_se1=SE1.Aligned.out.bam\n",
    "bam_se2=SE2.Aligned.out.bam\n",
    "\n",
    "\n",
    "\n",
    "# run STAR mapping ------------------—------------------—-----------------------\n",
    "\n",
    "cd $projdir/mapping_star\n",
    "\n",
    "# load genome index [5~10 min]\n",
    "# creates some apparent .log, .sam out despite just loading genome\n",
    "# so putting in mapping_star to keep these files in one place\n",
    "\n",
    "STAR --runThreadN 4 --genomeDir $ref_starfolder --genomeLoad LoadAndExit\n",
    "\n",
    "# map each well\n",
    "for row in ${target_well_rows[@]} \n",
    "do\n",
    "\n",
    "    # check directory/prior mapping ------------------—------------------—----\n",
    "    \n",
    "    cd $projdir\n",
    "\n",
    "    # check for existing mapping output\n",
    "    # if final outputs exist, skip; else run mapping .bam\n",
    "    \n",
    "    if [[ -s ${dir_well[$row]}/$bam_pe \\\n",
    "        && -s ${dir_well[$row]}/$bam_se1 \\\n",
    "        && -s ${dir_well[$row]}/$bam_se2 \\\n",
    "        && \"$skip_complete\"==\"true\" ]]\n",
    "    then\n",
    "        echo -e \"final aligned .bams for '${wellprefix[$row]}' already exist. skipping this well.'\"\n",
    "    else\n",
    "    \n",
    "    echo -e \"\\n\\napplying STAR to '${wellprefix[$row]}'...\\n\\n\"\n",
    "\n",
    "    # remove old directory if one exists to deal with incomplete files\n",
    "    # albeit the only major issues are .bai and .tbi indices \n",
    "    # (these often are not overwritten by software in the pipeline,\n",
    "    # resulting in \"index is older than file\" errors later on)\n",
    "    if [[ -e ${dir_well[$row]} && \"$overwrite_partial\" == \"true\" ]]\n",
    "    then\n",
    "        echo -e \"\\n\\nWARNING: folder for '${wellprefix[$row]}' exists, but not its final .bam alignments.\"\n",
    "        echo \"because overwrite_partial=true, deleting the directory and re-mapping.\"\n",
    "        rm -rf ${dir_well[$row]}\n",
    "    fi\n",
    "\n",
    "    mkdir ${dir_well[$row]}\n",
    "    cd ${dir_well[$row]}\n",
    "    \n",
    "    \n",
    "    # run alignments ------------------—------------------—-------------------\n",
    "    # in: .fastqs from trimming: four .fastqs,\n",
    "    #     properly paired ($fastq_r2p, $fastq_r1p) and trimming singletons ($fastq_r1singletrim, $fastq_r2singletrim)\n",
    "    # out: - paired-end, single-end .bam alignments out ($bam_pe, $bam_se1, $bam_se2)\n",
    "    #      - key log files (e.g., mapping rate) \n",
    "    # ---------------------------------------—------------------—-------------\n",
    "\n",
    "   # (i) paired-end mapping [<1-3 min] \n",
    "   # assumptions: pairs that map ambiguously in paired-end mode should be discarded\n",
    "    STAR $star_params \\\n",
    "        --outFileNamePrefix PE. \\\n",
    "        --readFilesIn $projdir/${fastq_r1p[$row]} $projdir/${fastq_r2p[$row]} \\\n",
    "        --outReadsUnmapped Fastx\n",
    "\n",
    "    # .fq --> .fq.gz for future storage/help match STAR's expected input type [<1-2 min]\n",
    "    bgzip $fastq_pe_unmap1\n",
    "    bgzip $fastq_pe_unmap2\n",
    "\n",
    "    # (ii.) single-end, R1 [<1-3 min]\n",
    "    # includes Read 1 singletons from trimming and STAR mapping in (i)\n",
    "    STAR $star_params \\\n",
    "        --outFileNamePrefix SE1. \\\n",
    "        --readFilesIn $projdir/${fastq_r1singletrim[$row]},$fastq_pe_unmap1.gz\n",
    "\n",
    "    # (iii.) single-end, R2 [<1-3 min]\n",
    "    # includes Read 2 singletons from trimming and STAR mapping in (i)\n",
    "    STAR $star_params \\\n",
    "        --outFileNamePrefix SE2. \\\n",
    "        --readFilesIn $projdir/${fastq_r2singletrim[$row]},$fastq_pe_unmap2.gz\n",
    "\n",
    "    # (iv.) optional clean-up (comment out as desired)\n",
    "    # *.Log.final.out contains mapping rate, whereas other *.out have STAR internals\n",
    "    # *.SJ.out.tab contain splice junctions\n",
    "    rm -rf *_STARtmp\n",
    "    rm *Log.progress.out\n",
    "    rm *Log.out\n",
    "\n",
    "fi\n",
    "done\n",
    "\n",
    "\n",
    "# unload at end\n",
    "STAR --genomeDir $refstar --genomeLoad Remove\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A05a_star_mapping' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `date `\n",
    "echo \" \""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A05b_check_star.sub\n",
    "\n",
    "# #!/bin/bash\n",
    "# #$ -cwd\n",
    "# #$ -o sublogs/A05b_check_star.$JOB_ID\n",
    "# #$ -j y\n",
    "# #$ -l h_rt=2:00:00,h_data=4G\n",
    "# #$ -N A05b_starcheck\n",
    "# #$ -hold_jid A05a_star\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init ------------------—------------------—-----------------------\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths ------------------—------------------—---------------\n",
    "\n",
    "cd $projdir\n",
    "\n",
    "query_metadat () {\n",
    "  awk -F',' -v targetcol=\"$1\" \\\n",
    "      'NR==1 {\n",
    "                for (i=1;i<=NF;i++) {\n",
    "                    if ($i==targetcol) {assayout=i; break} }\n",
    "                print $assayout\n",
    "              } \n",
    "      NR>1 {\n",
    "                print $assayout\n",
    "            }' $metadat_well\n",
    "}\n",
    "\n",
    "check_filepaths_in_assay() {\n",
    "    for file in $@\n",
    "        do \n",
    "        if [[ ! -s $file ]]\n",
    "            then\n",
    "                echo \"missing '$file'\"\n",
    "            fi\n",
    "        done\n",
    "}\n",
    "\n",
    "check_filepath_in_batch() {\n",
    "target_array=$@\n",
    "for ((target_batch=1; target_batch<=$nbatches; target_batch++))\n",
    "    do\n",
    "        target_well_rows=()\n",
    "        for ((row=1; row<=nwells; row++))\n",
    "        do\n",
    "            if [[ \"${batchnum[$row]}\" == \"$target_batch\" ]]\n",
    "            then\n",
    "                target_well_rows+=($row)\n",
    "            fi\n",
    "        done\n",
    "        \n",
    "        batch_allc_list=()\n",
    "        for row in ${target_well_rows[@]}\n",
    "        do\n",
    "            batch_allc_list+=(${allctbi[$row]})\n",
    "        done\n",
    "        \n",
    "        num_allc_missing=$(check_filepaths_in_assay ${batch_allc_list[@]} | wc -l)\n",
    "        echo -e \"${target_batch} \\t $num_allc_missing\"\n",
    "\n",
    "    done \\\n",
    "    | awk -F \"\\t\" '($2 > 0) { print $0 }'\n",
    "}\n",
    "\n",
    "batchnum=($(query_metadat \"batchnum\"))\n",
    "\n",
    "nwells=${#batchnum[@]}\n",
    "nbatches=${batchnum[-1]}\n",
    "\n",
    "\n",
    "\n",
    "# apply checks for A05a output -------—------------------—----------------------\n",
    "\n",
    "echo \"-----------------------------------------------------------------\"\n",
    "echo \"A. printing number of missing .bam files missing (by batch)... \"\n",
    "echo \"-----------------------------------------------------------------\"\n",
    "\n",
    "bam_star_pe=($(query_metadat \"A05a_bam_star_PE\"))\n",
    "bam_star_se1=($(query_metadat \"A05a_bam_star_SE1\"))\n",
    "bam_star_se2=($(query_metadat \"A05a_bam_star_SE2\"))\n",
    "\n",
    "echo \"checking PE.Aligned.out.bam\"\n",
    "echo -e \"batchnum\\tnum_missing\"\n",
    "check_filepath_in_batch ${bam_star_pe[@]}\n",
    "\n",
    "echo \"checking SE1.Aligned.out.bam\"\n",
    "echo -e \"batchnum\\tnum_missing\"\n",
    "check_filepath_in_batch ${bam_star_se1[@]}\n",
    "\n",
    "echo \"checking SE2.Aligned.out.bam\"\n",
    "echo -e \"batchnum\\tnum_missing\"\n",
    "check_filepath_in_batch ${bam_star_se2[@]}\n",
    "\n",
    "echo -e \"\\n\\nsuggest re-running and checking sublog output of above batches.\"\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n-----------------------------------------------------------------\"\n",
    "echo \"B. checking log files for issues.\"\n",
    "echo -e \"-----------------------------------------------------------------\\n\"\n",
    "\n",
    "\n",
    "echo \"checking if 'completed' in sublogs/A05a_star* output.\"\n",
    "echo \"if any filename is printed, the associated batch may have not completed mapping.\"\n",
    "\n",
    "grep -c 'A05a_star_mapping completed' sublogs/A05a_star* | awk -F \":\" '$2==0 {print $1}'\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A05b_starcheck' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID ended on:   \" `date `\n",
    "echo \" \""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A05c_classify_mCT_reads_STAR.pl\n",
    "\n",
    "#!/usr/bin/perl -w\n",
    "use strict;\n",
    "\n",
    "\n",
    "# A05c_classify_mCT_reads_STAR.pl, v0.2 =====================================================\n",
    "# based loosely on original perl script written by Dr. Chongyuan Luo (@luogenomics)\n",
    "# modifications by Choo Liu (@chooliu):\n",
    "# - readability/documentation\n",
    "# - changes in logic for paired-end mapping (fwd/rev strand) & more efficient parsing of MD:Z flag\n",
    "#   incl looking at only C/G positions in read vs. ref changes, calculating # cytosines, etc\n",
    "# inputs: - .sam file from STAR (compatible with single-end or paired-end alignments)\n",
    "# outputs: - \"_annotations\" .tsv recording each alignment's:\n",
    "#             number of cytosines, mC/C fraction, and call (DNA, RNA, ambiguous)\n",
    "#             this annotation file is subsequently appended to the. bam to keep RNA reads\n",
    "# typical usage: perl A05c_classify_mCT_reads_STAR.pl alignments.sam\n",
    "# =====================================================================================\n",
    "\n",
    "# reading in .sam file\n",
    "my ($samfile)=($ARGV[0]);\n",
    "my @sample; my @samline; \n",
    "my $read; my @read; my @ref; my $ref; my @md; my $dir; my $unmch; \n",
    "\n",
    "# determining read classification\n",
    "my $filter_num_CH=3;\n",
    "my $filter_frac_mCHmin=0.5; # DNA (mCH/CH <0.5)\n",
    "my $filter_frac_mCHmax=0.9; # RNA (mCH/CH >0.9)\n",
    "my $mch_fraction; my $totalch; my $totcpg; my @call;\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# load .sam file, loop through lines in .sam \n",
    "# exports info on each read to \"(input .sam file name)_annotations\"\n",
    "\n",
    "# for each line in .sam file, --------------------------------------------------------\n",
    "# count unmethylated and methylated cytosines in CHG and CHH context based on XR-tag \n",
    "\n",
    "# notes:\n",
    "# - skip header header rows (starting with @)\n",
    "# - assumes the XR-tag is in column 10 (STAR output), $samline[9]\n",
    "# - currently ignores indels\n",
    "# - default settings: keep reads with >=3 CHNs and mCHN/CHN fraction >=0.9 [*]\n",
    "#   at present, we only retain call=\"RNA\" and discard all others\n",
    "\n",
    "open sam_in, \"$samfile\" or die $!;\n",
    "open sam_annotations, \">$samfile\\_annotations\" or die $!;\n",
    "while (<sam_in>)\n",
    "{\n",
    "  chop $_; \n",
    "  if (substr($_,0,1) eq '@') { print sam_annotations \"\\n\"; }\n",
    "  else\n",
    "  {\n",
    "  \n",
    "    @samline = split(/\\t/,$_);\n",
    "    $read = $samline[9];\n",
    "    @read = split(//,$read);\n",
    "    @ref = @read;\n",
    "    \n",
    "    @md=split( /(\\d+)/ , $samline[15]);\n",
    "    \n",
    "    # check mapping orientation relative to reference genome ----------------------------\n",
    "    # dir=1 if fwd (99, 163, 0, 73), =0 if reverse (147, 83, 16, 89)\n",
    "    # flaw in script is that these numbers manually specified--if see strange results,\n",
    "    # check mapping output for other flags before filtering / add interpreter\n",
    "    $dir = 0 + ( ($samline[1] eq 99) or ($samline[1] eq 163 ) or ($samline[1] eq 0) or ($samline[1] eq 73) ); \n",
    "    \n",
    "    # compare read to reference genome --------------------------------------------------\n",
    "    # split MD:Z: flag by numbers then examine resulting length\n",
    "    # e.g., MD:Z:10A0B121 --> MD:Z: 10 A 0 B 121 has $num_md_features=5\n",
    "    \n",
    "    # assume fully methylated if perfect match to genomic ref\n",
    "    # (if MD stores a single number indicating no changes, $#md = 1)\n",
    "    my $unmch = 0;\n",
    "    my $num_md_features=$#md;\n",
    "    if ($num_md_features == 1) { }\n",
    "    \n",
    "    # <-- start ELSE for num_md_features != 1\n",
    "    # otherwise, attempt to reconstruct reference sequence, \n",
    "    # looping through positions in read where the read != ref nucleotide\n",
    "    else {\n",
    "        my $pos=0;\n",
    "\n",
    "        for (my $i=1; $i<=$num_md_features; $i += 2) {\n",
    "        \n",
    "            # loop through bases where there are changes\n",
    "            $pos = $pos + @md[$i];\n",
    "            \n",
    "            # do nothing if starts with \"^\" (deletion)\n",
    "            # modify @ref sequence if differences\n",
    "             if ( (rindex(\"@md[$i+1]\", \"^\", 0) == 0) or ($i+1 > $num_md_features)) { }  \n",
    "             else {\n",
    "                 @ref[$pos] = @md[$i + 1];\n",
    "                 }\n",
    "            \n",
    "            # increment base by 1\n",
    "            $pos += 1;\n",
    "        }\n",
    "  \n",
    "    # tabulate # unmethylated cytosines --------------------------------------------------- \n",
    "    # again looping through each position where read != ref\n",
    "    # (although same loop as above, re-loop because potential cases where adjacent bases differ btwn ref & read)\n",
    "        my $pos=0;\n",
    "        for (my $i=1; $i<=$num_md_features; $i += 2) {\n",
    "        \n",
    "           $pos = $pos + @md[$i];\n",
    "           \n",
    "           # if read maps to forward strand of genome\n",
    "           # check if at a CH-site, and unmethylated cytosine converted to \"T\"\n",
    "           if ( ($dir eq 1) and\n",
    "                (@ref[$pos] eq \"C\") and (@ref[$pos + 1] ne \"G\") and (@read[$pos] eq \"T\") ) {\n",
    "                    $unmch += 1;\n",
    "                }\n",
    "       \n",
    "           # if read maps to reverse strand of genome\n",
    "           # STAR rev compliments the $read sequence, so cytosines are represented by \"G\"\n",
    "           # check if at a CH-site, and unmethylated cytosine converted to \"A\"\n",
    "           # (note: mCH underestimated in niche case where dir=0 & \n",
    "           #        first base is cytosine, as @ref[$pos - 1] is undefined)\n",
    "           if ( ($dir eq 0) and\n",
    "                (@ref[$pos] eq \"G\") and (@ref[$pos - 1] ne \"C\") and (@read[$pos] eq \"A\") ) {\n",
    "                $unmch += 1;\n",
    "            }\n",
    "            \n",
    "            $pos += 1;\n",
    "        }\n",
    "    } # <-- end ELSE statement for num_md_features != 1\n",
    "\n",
    "    $ref = join('', @ref);\n",
    "    \n",
    "    # count cytosines in CH-context --------------------------------------------------\n",
    "    # note: since done via regex, faster to count [# C] and subtract [# CG] vs searching wild flag C[ACT]\n",
    "    if ($dir eq 1) {\n",
    "      my @totalch = $ref =~ /C/g;\n",
    "      $totalch = scalar @totalch;\n",
    "      my @totcpg = $ref =~ /CG/g;\n",
    "      $totcpg = scalar @totcpg;\n",
    "    }\n",
    "    if ($dir eq 0) {\n",
    "      my @totalch = $ref =~ /G/g;\n",
    "      $totalch = scalar @totalch;\n",
    "      my @totcpg = $ref =~ /CG/g;\n",
    "      $totcpg = scalar @totcpg;\n",
    "    }\n",
    "\n",
    "    my $totalch = $totalch - $totcpg;\n",
    "    \n",
    "    # classify each read into modalities ----------------------------------------------\n",
    "    if ($totalch==0) { # avoid div by zero error\n",
    "          $mch_fraction=-999;\n",
    "          @call=\"amb\";\n",
    "    } else {\n",
    "          $mch_fraction = 1 - ($unmch/$totalch);\n",
    "          if (($totalch>=$filter_num_CH) and ($mch_fraction < $filter_frac_mCHmin)) { @call=\"DNA\"; }\n",
    "          elsif (($totalch>=$filter_num_CH) and ($mch_fraction >= $filter_frac_mCHmax)) { @call=\"RNA\"; }\n",
    "          else { @call=\"amb\"; } # exclude (low # CH or ambiguous mCH between 0.5-0.9)\n",
    "    }\n",
    "\n",
    "    print sam_annotations \"${totalch}\\t${mch_fraction}\\t@call\\n\";\n",
    "\n",
    "}\n",
    "}\n",
    "\n",
    "close bam_in;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A05c_star_filtering.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A05c_starfilt.$JOB_ID.$TASK_ID\n",
    "#$ -j y\n",
    "#$ -l h_rt=8:00:00,h_data=8G\n",
    "#$ -N A05c_starfilt\n",
    "#$ -t 1-256\n",
    "#$ -hold_jid_ad A05a_star\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init ------------------—------------------—-----------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snmCTseq # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "skip_complete=true # <-- for help with incomplete jobs\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths ------------------—------------------—---------------\n",
    "\n",
    "# helper functions\n",
    "query_metadat () {\n",
    "  awk -F',' -v targetcol=\"$1\" \\\n",
    "      'NR==1 {\n",
    "                for (i=1;i<=NF;i++) {\n",
    "                    if ($i==targetcol) {assayout=i; break} }\n",
    "                print $assayout\n",
    "              }\n",
    "      NR>1 {\n",
    "                print $assayout\n",
    "            }' $metadat_well\n",
    "}\n",
    "\n",
    "# extract target wells, print values for log\n",
    "\n",
    "batchnum=($(query_metadat \"batchnum\"))\n",
    "nwells=${#batchnum[@]}\n",
    "\n",
    "target_well_rows=()\n",
    "for ((row=1; row<=nwells; row++))\n",
    "do\n",
    "    if [[ \"${batchnum[$row]}\" == \"$SGE_TASK_ID\" ]]\n",
    "    then\n",
    "        target_well_rows+=($row)\n",
    "    fi\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "# filepaths associated with target rows in well-level metadata -----------------\n",
    "\n",
    "wellprefix=($(query_metadat \"wellprefix\"))\n",
    "dir_well=($(query_metadat \"A05a_dir_star\"))\n",
    "\n",
    "\n",
    "\n",
    "# set naming convention within each well folder -------------—------------------\n",
    "\n",
    "# .bam files in (from A05a)\n",
    "bam_in_pe=PE.Aligned.out.bam\n",
    "bam_in_se1=SE1.Aligned.out.bam\n",
    "bam_in_se2=SE2.Aligned.out.bam\n",
    "\n",
    "# final .sam\n",
    "sam_q10_pe=q10_pe.sam\n",
    "sam_q10_se1=q10_se1.sam\n",
    "sam_q10_se2=q10_se2.sam\n",
    "\n",
    "# final .bam\n",
    "bam_final_pe=PE.Final.bam\n",
    "bam_final_se1=SE1.Final.bam\n",
    "bam_final_se2=SE2.Final.bam\n",
    "\n",
    "# map each well\n",
    "for row in ${target_well_rows[@]} \n",
    "do\n",
    "\n",
    "    # check directory/prior mapping ------------------—------------------—-----\n",
    "    # check for existing mapping output\n",
    "    # if final outputs exist, skip; else run filtering of direct STAR align\n",
    "    \n",
    "    cd $projdir\n",
    "    \n",
    "    if [[ -s ${dir_well[$row]}/$bam_final_pe \\\n",
    "        && -s ${dir_well[$row]}/$bam_final_se1 \\\n",
    "        && -s ${dir_well[$row]}/$bam_final_se2 \\\n",
    "        && \"$skip_complete\"==\"true\" ]]\n",
    "    then\n",
    "        echo -e \"final aligned .bams for '${wellprefix[$row]}' already exist. skipping this well.'\"\n",
    "    else\n",
    "    \n",
    "        echo -e \"\\n\\napplying STAR to '${wellprefix[$row]}'...\\n\\n\"\n",
    "        cd ${dir_well[$row]}\n",
    "        \n",
    "\n",
    "        # proceed if all input files exist\n",
    "        if [[ -s $bam_in_pe && -s $bam_in_se1 && -s $bam_in_se2 ]]\n",
    "        then\n",
    "\n",
    "            # clear intermediate files (if not all of them exist)\n",
    "            rm $bam_final_pe $bam_final_se1 $bam_final_se2 2>&1 >/dev/null\n",
    "\n",
    "        # run RNA filtering ------------------—------------------—-----------------\n",
    "        # in: three .bams from STAR mapping: $bam_in_* (paired-end, single-end read 1, read 2)\n",
    "        # out: - MAPQ and f ($bam_final_*)\n",
    "        #      - key log files (e.g., mapping rate) \n",
    "        # ------------------—------------------—-----------------------------------\n",
    "        \n",
    "        # STAR will output singletons in .bam file\n",
    "        # (e.g., read 2 maps but read 1 doesn't; samtools view -f 8 $bam_in_pe)\n",
    "        # hence the \"-f 0x0002\" flag for \"proper pairs\" only, and SE alignments merged with...\n",
    "        # in later sections, \"-f 0x0048\" (the read is R1; it mapped but R2 mate didn't map) \n",
    "        #                and \"-f 0x0088\" (the read is R2; it mapped but R1 mate didn't map)\n",
    "\n",
    "        # if both R1 & R2 of pair need to pass filtering criteria (AND instead of OR),\n",
    "        #     needs an added annotations --> wide step like the below\n",
    "        # sed '$!N;s/\\n/ /' ${sam_q10_pe}_annotations \\\n",
    "        #     | awk '{print ( ($3 == \"RNA\") && ($6 == \"RNA\") )}' \\\n",
    "        #     | awk '{print $0}1' > ${sam_q10_pe}_annotations_bothpairs\n",
    "        # then awk filter off of this 'bothpairs' file\n",
    "        \n",
    "        # each step usually ~1-2 min\n",
    "\n",
    "        # (i) paired-end\n",
    "        samtools view -h -q 10 -f 0x0002 $bam_in_pe > $sam_q10_pe\n",
    "        perl $projdir/Scripts/A05c_classify_mCT_reads_STAR.pl $sam_q10_pe\n",
    "        awk 'NR == FNR { if ($0==\"\" || $3==\"RNA\") line[NR]; next } (FNR in line)' \\\n",
    "                ${sam_q10_pe}_annotations ${sam_q10_pe} |\n",
    "                samtools view -b - | samtools sort - > $bam_final_pe\n",
    "        samtools index $bam_final_pe\n",
    "\n",
    "        # (ii) single-end, read 1\n",
    "        samtools view -h -q 10 $bam_in_se1 > $sam_q10_se1\n",
    "        perl $projdir/Scripts/A05c_classify_mCT_reads_STAR.pl $sam_q10_se1\n",
    "        awk 'NR == FNR { if ($0==\"\" || $3==\"RNA\") line[NR]; next } (FNR in line)' \\\n",
    "                ${sam_q10_se1}_annotations ${sam_q10_se1} |\n",
    "                samtools view -b - | samtools sort - > $bam_final_se1\n",
    "        samtools index $bam_final_se1\n",
    "\n",
    "        # (iii) single-end, read 2\n",
    "        samtools view -h -q 10 $bam_in_se2 > $sam_q10_se2\n",
    "        perl $projdir/Scripts/A05c_classify_mCT_reads_STAR.pl $sam_q10_se2\n",
    "        awk 'NR == FNR { if ($0==\"\" || $3==\"RNA\") line[NR]; next } (FNR in line)' \\\n",
    "                ${sam_q10_se2}_annotations ${sam_q10_se2} |\n",
    "                samtools view -b - | samtools sort - > $bam_final_se2\n",
    "        samtools index $bam_final_se2\n",
    "\n",
    "        # (iv.) optional clean-up (comment out as desired)\n",
    "        rm $sam_q10_pe $sam_q10_se1 $sam_q10_se2\n",
    "        # rm *annotations\n",
    "\n",
    "        fi\n",
    "    fi\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "echo \"completed 'A05c_star_filtering.'\"\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `date `\n",
    "echo \" \""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A05d_featurecounts.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A05d_featurecounts.$JOB_ID.$TASK_ID\n",
    "#$ -j y\n",
    "#$ -l h_rt=16:00:00,h_data=24G\n",
    "#$ -N A05d_featurecounts\n",
    "#$ -t 1-16\n",
    "#$ -hold_jid A05c_starfilt\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "# environment init ------------------—------------------—-----------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snmCTseq # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "dir_out_gene=featurecounts_gene/ # <-- \n",
    "dir_out_exon=featurecounts_exon/ # <--\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths ------------------—------------------—---------------\n",
    "\n",
    "mkdir $dir_out_gene\n",
    "mkdir $dir_out_exon\n",
    "\n",
    "# helper functions\n",
    "query_metadat () {\n",
    "  awk -F',' -v targetcol=\"$1\" \\\n",
    "      'NR==1 {\n",
    "                for (i=1;i<=NF;i++) {\n",
    "                    if ($i==targetcol) {assayout=i; break} }\n",
    "                print $assayout\n",
    "              }\n",
    "      NR>1 {f\n",
    "                print $assayout\n",
    "            }' $metadat_well\n",
    "}\n",
    "\n",
    "# extract target wells, print values for log\n",
    "\n",
    "platenum=($(query_metadat \"platenum\"))\n",
    "nwells=${#platenum[@]}\n",
    "\n",
    "target_well_rows=()\n",
    "for ((row=1; row<=nwells; row++))\n",
    "do\n",
    "    if [[ \"${platenum[$row]}\" == \"$SGE_TASK_ID\" ]]\n",
    "    then\n",
    "        target_well_rows+=($row)\n",
    "    fi\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "# filepaths associated with target rows in well-level metadata -----------------\n",
    "\n",
    "wellprefix=($(query_metadat \"wellprefix\"))\n",
    "dir_well=($(query_metadat \"A05a_dir_star\"))\n",
    "\n",
    "bam_pe=($(query_metadat \"A05c_bam_starfilt_PE\"))\n",
    "bam_se1=($(query_metadat \"A05c_bam_starfilt_SE1\"))\n",
    "bam_se2=($(query_metadat \"A05c_bam_starfilt_SE2\"))\n",
    "\n",
    "\n",
    "\n",
    "# extract valid .bam filepaths -------------------------------------------------\n",
    "\n",
    "# checks only if the .bam exists (otherwise featureCounts may terminate with error)\n",
    "# could also wrap in check that number of alignments in file is > 0\n",
    "# for greater future compatibility with featureCounts\n",
    "\n",
    "pe_files=$(\n",
    "for f in ${bam_pe[@]: ${target_well_rows[0] }:${#target_well_rows[@] } }\n",
    "do\n",
    "    if [[ -s $f ]]\n",
    "    then\n",
    "        echo $f\n",
    "    fi\n",
    "done)\n",
    "\n",
    "se1_files=$(\n",
    "for f in ${bam_se1[@]: ${target_well_rows[0] }:${#target_well_rows[@] } }\n",
    "do\n",
    "    if [[ -s $f ]]\n",
    "    then\n",
    "        echo $f\n",
    "    fi\n",
    "done)\n",
    "\n",
    "se2_files=$(\n",
    "for f in ${bam_se2[@]: ${target_well_rows[0] }:${#target_well_rows[@] } }\n",
    "do\n",
    "    if [[ -s $f ]]\n",
    "    then\n",
    "        echo $f\n",
    "    fi\n",
    "done)\n",
    "\n",
    "\n",
    "\n",
    "# featurecounts on genes ------------------—------------------—-----------------\n",
    "# usually <1 min/well --> ~6 hrs per plate\n",
    "\n",
    "echo \"running gene featureCounts on paired-end alignments (mapping_star/*/PE.Final.bam).\"\n",
    "featureCounts -p -T 4 -t gene -a $ref_gtf \\\n",
    "    -o ${dir_out_gene}PE_${SGE_TASK_ID} --donotsort --tmpDir $dir_scratch $pe_files\n",
    "\n",
    "echo \"running gene featureCounts on single-end R1 alignments (mapping_star/*/SE1.Final.bam).\"\n",
    "se1_files=${bam_se1[@]: ${target_well_rows[0] }:${#target_well_rows[@] } }\n",
    "featureCounts -T 4 -t gene -a $ref_gtf \\\n",
    "    -o ${dir_out_gene}SE1_${SGE_TASK_ID} --donotsort --tmpDir $dir_scratch $se1_files\n",
    "\n",
    "echo \"running gene featureCounts on single-end R2 alignments (mapping_star/*/SE2.Final.bam).\"\n",
    "se2_files=${bam_se2[@]: ${target_well_rows[0] }:${#target_well_rows[@] } }\n",
    "featureCounts -T 4 -t gene -a $ref_gtf \\\n",
    "    -o ${dir_out_gene}SE2_${SGE_TASK_ID} --donotsort --tmpDir $dir_scratch $se2_files\n",
    "\n",
    "\n",
    "\n",
    "# featurecounts on exons ------------------—------------------—-----------------\n",
    "\n",
    "echo \"running exon featureCounts on paired-end alignments (mapping_star/*/PE.Final.bam).\"\n",
    "featureCounts -p -T 4 -t exon -a $ref_gtf \\\n",
    "    -o ${dir_out_exon}PE_${SGE_TASK_ID} --donotsort --tmpDir $dir_scratch $pe_files\n",
    "\n",
    "echo \"running exon featureCounts on single-end R1 alignments (mapping_star/*/SE1.Final.bam).\"\n",
    "featureCounts -T 4 -t exon -a $ref_gtf \\\n",
    "    -o ${dir_out_exon}SE1_${SGE_TASK_ID} --donotsort --tmpDir $dir_scratch $se1_files\n",
    "\n",
    "echo \"running exon featureCounts on single-end R2 alignments (mapping_star/*/SE2.Final.bam).\"\n",
    "featureCounts -T 4 -t exon -a $ref_gtf \\\n",
    "    -o ${dir_out_exon}SE2_${SGE_TASK_ID} --donotsort --tmpDir $dir_scratch $se2_files\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A05d_featurecounts' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A05e_star_bam_stats.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A05g_samstat_star.$JOB_ID.$TASK_ID\n",
    "#$ -j y\n",
    "#$ -l h_rt=8:00:00,h_data=24G\n",
    "#$ -N A05e_samstat_star\n",
    "#$ -t 1-256\n",
    "#$ -hold_jid_ad A05c_starfilt\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "# environment init ------------------—------------------—-----------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snmCTseq # <--\n",
    "\n",
    "export $(cat snmCT_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "skip_complete=true # <-- for help with incomplete jobs\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths ------------------—------------------—---------------\n",
    "\n",
    "# helper functions\n",
    "query_metadat () {\n",
    "  awk -F',' -v targetcol=\"$1\" \\\n",
    "      'NR==1 {\n",
    "                for (i=1;i<=NF;i++) {\n",
    "                    if ($i==targetcol) {assayout=i; break} }\n",
    "                print $assayout\n",
    "              }\n",
    "      NR>1 {\n",
    "                print $assayout\n",
    "            }' $metadat_well\n",
    "}\n",
    "\n",
    "# extract target wells, print values for log\n",
    "\n",
    "batchnum=($(query_metadat \"batchnum\"))\n",
    "nwells=${#batchnum[@]}\n",
    "\n",
    "target_well_rows=()\n",
    "for ((row=1; row<=nwells; row++))\n",
    "do\n",
    "    if [[ \"${batchnum[$row]}\" == \"$SGE_TASK_ID\" ]]\n",
    "    then\n",
    "        target_well_rows+=($row)\n",
    "    fi\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "# filepaths associated with target rows in well-level metadata -----------------\n",
    "\n",
    "wellprefix=($(query_metadat \"wellprefix\"))\n",
    "dir_well=($(query_metadat \"A05a_dir_star\"))\n",
    "\n",
    "\n",
    "\n",
    "# samtools stats on each well in the batch --—------------------—---------------\n",
    "\n",
    "for row in ${target_well_rows[@]} \n",
    "do\n",
    "\n",
    "    cd $projdir\n",
    "    \n",
    "    if [[ -s ${dir_well[$row]}/samstats_PE \\\n",
    "        && -s ${dir_well[$row]}/samstats_SE1 \\\n",
    "        && -s ${dir_well[$row]}/samstats_SE2 \\\n",
    "        && -s ${dir_well[$row]}/picard_PE \\\n",
    "        && -s ${dir_well[$row]}/picard_SE1 \\\n",
    "        && -s ${dir_well[$row]}/picard_SE2 \\\n",
    "        && \"$skip_complete\"==\"true\" ]]\n",
    "    then\n",
    "        echo -e \"final aligned .bams for '${wellprefix[$row]}' already exist. skipping this well.'\"\n",
    "    else\n",
    "    \n",
    "        echo -e \"\\n\\nprofiling .bams from '${wellprefix[$row]}'...\\n\\n\"\n",
    "\n",
    "        cd ${dir_well[$row]}\n",
    "        \n",
    "        # run samtools stats\n",
    "        samtools stats PE.Final.bam | grep '^SN' | cut -f 2,3 > samstats_PE\n",
    "        samtools stats SE1.Final.bam | grep '^SN' | cut -f 2,3 > samstats_SE1\n",
    "        samtools stats SE2.Final.bam | grep '^SN' | cut -f 2,3 > samstats_SE2\n",
    "\n",
    "        # run picard collectrnaseqmetrics\n",
    "        samtools sort PE.Final.bam | \\\n",
    "            picard CollectRnaSeqMetrics -I /dev/stdin -O picard_PE \\\n",
    "            --REF_FLAT $ref_flat -STRAND \"NONE\" --RIBOSOMAL_INTERVALS $ref_rrna\n",
    "        samtools sort SE1.Final.bam | \\\n",
    "            picard CollectRnaSeqMetrics -I /dev/stdin -O picard_SE1 \\\n",
    "            --REF_FLAT $ref_flat -STRAND \"NONE\" --RIBOSOMAL_INTERVALS $ref_rrna\n",
    "        samtools sort SE2.Final.bam | \\\n",
    "            picard CollectRnaSeqMetrics -I /dev/stdin -O picard_SE2 \\\n",
    "            --REF_FLAT $ref_flat -STRAND \"NONE\" --RIBOSOMAL_INTERVALS $ref_rrna\n",
    "            \n",
    "    fi\n",
    "done\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A05e_samstat_star' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `date `\n",
    "echo \" \"\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
